{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## effect size sharing for multi-gene qtls\n",
    "\n",
    "say we have a qtl that impacts multiple genes. How does the fractional effect of that qtl on each gene vary? w.r.t for instance distance from qtl to tss? maybe tie in ABC connections here? Maybe build a model that predicts how qtl impact on each gene will be split based on gene baseline expression, distance, ect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt \n",
    "import seaborn as sns \n",
    "import os\n",
    "import ast\n",
    "import re\n",
    "\n",
    "import sys\n",
    "sys.path.append('/home/klawren/oak/pcqtls/workflow/scripts')\n",
    "from notebook_helper_functions import *\n",
    "\n",
    "import upsetplot as up\n",
    "from tqdm.auto import tqdm  # for notebooks\n",
    "\n",
    "# Create new `pandas` methods which use `tqdm` progress\n",
    "# (can use tqdm_gui, optional kwargs, etc.)\n",
    "tqdm.pandas()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get outputs from a config file\n",
    "prefix = '/home/klawren/oak/pcqtls'\n",
    "import yaml\n",
    "config_path= f'{prefix}/config/proteincoding_main.yaml'\n",
    "with open(config_path, 'r') as f:\n",
    "    config = yaml.safe_load(f)\n",
    "\n",
    "# load in the tissue ids \n",
    "tissue_df = load_tissue_df(config)\n",
    "tissue_ids = list(tissue_df['Tissue'])\n",
    "\n",
    "# select just 1 tissue id to do for now \n",
    "my_tissue_id = 'Cells_Cultured_fibroblasts' # because only one cell type\n",
    "#tissue_ids = [tissue_id]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### load in data\n",
    "effect sizes for qtls\n",
    "overlap dfs\n",
    "gencode for tss starts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "e_nominal = load_e_nominal(config, my_tissue_id, chr_id=22)\n",
    "pc_nominal = load_pc_nominal(config, my_tissue_id, chr_id=22)\n",
    "overlap = load_overlap(config, my_tissue_id)\n",
    "pc_susie = load_pc_susie(config, my_tissue_id)\n",
    "\n",
    "# set e_nominal to be index based on variant-cluster pairs\n",
    "e_nominal_cid = e_nominal.set_index('var_cluster')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "# look at just one chr to start\n",
    "pc_susie['chr'] = pc_susie['variant_id'].str.split('chr').str[1].str.split('_').str[0]\n",
    "pc_susie = pc_susie[pc_susie['chr'] == '22']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "take the subset of varaints in pcqtl determined credible sets, and calculate the varience-explained on a per gene basis for each gene in the cluster and for the cluster-pc overall\n",
    "\n",
    "variance for each egene-cluster-credibleset(pc and cs num)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'chr22_18933556_A_G_b38_ENSG00000100033.16_ENSG00000183628.12'"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pc_susie['var_cluster'].iloc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/local/scratch/klawren/slrmtmp.44431414/ipykernel_1284/2850190003.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  e_nominal_pcsusie_subset['var_cluster_egene'] = e_nominal_pcsusie_subset['variant_id'] + '_' + e_nominal_pcsusie_subset['phenotype_id']\n"
     ]
    }
   ],
   "source": [
    "# get a subset of variants that ended up in credible sets\n",
    "e_nominal_pcsusie_subset = e_nominal[e_nominal['var_cluster'].isin(pc_susie['var_cluster'])]\n",
    "\n",
    "# I want to get a pip weighted variance for each egene\n",
    "e_nominal_pcsusie_subset['var_cluster_egene'] = e_nominal_pcsusie_subset['variant_id'] + '_' + e_nominal_pcsusie_subset['phenotype_id']\n",
    "\n",
    "# merge in the data from the susie finemapping\n",
    "e_nominal_pcsusie_subset = pd.merge(e_nominal_pcsusie_subset, pc_susie[['cs_id', 'var_cluster', 'pip', 'pc_num', 'lead_variant_id', 'num_vars']], on='var_cluster', how='right')\n",
    "\n",
    "# pip weighted variance\n",
    "e_nominal_pcsusie_subset['variance'] = e_nominal_pcsusie_subset['slope'].apply(np.square) * 100\n",
    "e_nominal_pcsusie_subset['variance_weighted'] = e_nominal_pcsusie_subset['variance'] * e_nominal_pcsusie_subset['pip']\n",
    "\n",
    "# we should group based on unique sets of variants belonging to a certain cluster, egene, and credible set\n",
    "e_nominal_pcsusie_subset['cluster_egene_credibleset'] = e_nominal_pcsusie_subset['phenotype_id'] + '_cs' + '_pc' + e_nominal_pcsusie_subset['cs_id'].str.split('_pc').str[-1]\n",
    "\n",
    "# group by phenotype (egene and cluster)\n",
    "pcphenotype_evariance = e_nominal_pcsusie_subset.groupby('cluster_egene_credibleset').agg({'variance_weighted':sum, \n",
    "                                                        'cluster_id':'first', \n",
    "                                                        'variant_id':list,\n",
    "                                                        'egene_id':'first', \n",
    "                                                        'cs_id':'first', \n",
    "                                                        'lead_variant_id':'first',\n",
    "                                                        'num_vars':'first',\n",
    "                                                        'pc_num':'first'})\n",
    "\n",
    "pcphenotype_evariance = pcphenotype_evariance.rename(columns={'variant_id':'variant_list'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "phenotype_id                                                               NaN\n",
       "variant_id                                                                 NaN\n",
       "start_distance                                                             NaN\n",
       "end_distance                                                               NaN\n",
       "af                                                                         NaN\n",
       "ma_samples                                                                 NaN\n",
       "ma_count                                                                   NaN\n",
       "pval_nominal                                                               NaN\n",
       "slope                                                                      NaN\n",
       "slope_se                                                                   NaN\n",
       "cluster_id                                                                 NaN\n",
       "var_cluster                  chr22_20084214_T_C_b38_ENSG00000099904.15_ENSG...\n",
       "egene_id                                                                   NaN\n",
       "var_cluster_egene                                                          NaN\n",
       "cs_id                               ENSG00000099904.15_ENSG00000234409.6_pc2_1\n",
       "pip                                                                   0.066181\n",
       "pc_num                                                                       2\n",
       "lead_variant_id                                         chr22_20146827_C_A_b38\n",
       "num_vars                                                                    18\n",
       "variance                                                                   NaN\n",
       "variance_weighted                                                          NaN\n",
       "cluster_egene_credibleset                                                  NaN\n",
       "Name: 10, dtype: object"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "e_nominal_pcsusie_subset.iloc[10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variance_weighted</th>\n",
       "      <th>cluster_id</th>\n",
       "      <th>variant_list</th>\n",
       "      <th>egene_id</th>\n",
       "      <th>cs_id</th>\n",
       "      <th>lead_variant_id</th>\n",
       "      <th>num_vars</th>\n",
       "      <th>pc_num</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cluster_egene_credibleset</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [variance_weighted, cluster_id, variant_list, egene_id, cs_id, lead_variant_id, num_vars, pc_num]\n",
       "Index: []"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pcphenotype_evariance[pcphenotype_evariance['pc_num'] == 2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variance_weighted</th>\n",
       "      <th>cluster_id</th>\n",
       "      <th>variant_list</th>\n",
       "      <th>egene_id</th>\n",
       "      <th>cs_id</th>\n",
       "      <th>lead_variant_id</th>\n",
       "      <th>num_vars</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cluster_egene_credibleset</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000025708.13_ENSG00000177989.13_e_ENSG00000025708.13_cs_pc1_1</th>\n",
       "      <td>0.057016</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13</td>\n",
       "      <td>[chr22_50532837_T_C_b38]</td>\n",
       "      <td>ENSG00000025708.13</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_pc1_1</td>\n",
       "      <td>chr22_50532837_T_C_b38</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000025708.13_ENSG00000177989.13_e_ENSG00000025708.13_cs_pc1_2</th>\n",
       "      <td>0.189943</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13</td>\n",
       "      <td>[chr22_50504763_C_T_b38, chr22_50505337_G_A_b3...</td>\n",
       "      <td>ENSG00000025708.13</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_pc1_2</td>\n",
       "      <td>chr22_50519627_C_T_b38</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000025708.13_ENSG00000177989.13_e_ENSG00000177989.13_cs_pc1_1</th>\n",
       "      <td>0.023457</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13</td>\n",
       "      <td>[chr22_50532837_T_C_b38]</td>\n",
       "      <td>ENSG00000177989.13</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_pc1_1</td>\n",
       "      <td>chr22_50532837_T_C_b38</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000025708.13_ENSG00000177989.13_e_ENSG00000177989.13_cs_pc1_2</th>\n",
       "      <td>0.269531</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13</td>\n",
       "      <td>[chr22_50504763_C_T_b38, chr22_50505337_G_A_b3...</td>\n",
       "      <td>ENSG00000177989.13</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_pc1_2</td>\n",
       "      <td>chr22_50519627_C_T_b38</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000093000.18_ENSG00000100364.18_e_ENSG00000093000.18_cs_pc1_1</th>\n",
       "      <td>0.350770</td>\n",
       "      <td>ENSG00000093000.18_ENSG00000100364.18</td>\n",
       "      <td>[chr22_45105543_G_A_b38]</td>\n",
       "      <td>ENSG00000093000.18</td>\n",
       "      <td>ENSG00000093000.18_ENSG00000100364.18_pc1_1</td>\n",
       "      <td>chr22_45105543_G_A_b38</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    variance_weighted  \\\n",
       "cluster_egene_credibleset                                               \n",
       "ENSG00000025708.13_ENSG00000177989.13_e_ENSG000...           0.057016   \n",
       "ENSG00000025708.13_ENSG00000177989.13_e_ENSG000...           0.189943   \n",
       "ENSG00000025708.13_ENSG00000177989.13_e_ENSG000...           0.023457   \n",
       "ENSG00000025708.13_ENSG00000177989.13_e_ENSG000...           0.269531   \n",
       "ENSG00000093000.18_ENSG00000100364.18_e_ENSG000...           0.350770   \n",
       "\n",
       "                                                                               cluster_id  \\\n",
       "cluster_egene_credibleset                                                                   \n",
       "ENSG00000025708.13_ENSG00000177989.13_e_ENSG000...  ENSG00000025708.13_ENSG00000177989.13   \n",
       "ENSG00000025708.13_ENSG00000177989.13_e_ENSG000...  ENSG00000025708.13_ENSG00000177989.13   \n",
       "ENSG00000025708.13_ENSG00000177989.13_e_ENSG000...  ENSG00000025708.13_ENSG00000177989.13   \n",
       "ENSG00000025708.13_ENSG00000177989.13_e_ENSG000...  ENSG00000025708.13_ENSG00000177989.13   \n",
       "ENSG00000093000.18_ENSG00000100364.18_e_ENSG000...  ENSG00000093000.18_ENSG00000100364.18   \n",
       "\n",
       "                                                                                         variant_list  \\\n",
       "cluster_egene_credibleset                                                                               \n",
       "ENSG00000025708.13_ENSG00000177989.13_e_ENSG000...                           [chr22_50532837_T_C_b38]   \n",
       "ENSG00000025708.13_ENSG00000177989.13_e_ENSG000...  [chr22_50504763_C_T_b38, chr22_50505337_G_A_b3...   \n",
       "ENSG00000025708.13_ENSG00000177989.13_e_ENSG000...                           [chr22_50532837_T_C_b38]   \n",
       "ENSG00000025708.13_ENSG00000177989.13_e_ENSG000...  [chr22_50504763_C_T_b38, chr22_50505337_G_A_b3...   \n",
       "ENSG00000093000.18_ENSG00000100364.18_e_ENSG000...                           [chr22_45105543_G_A_b38]   \n",
       "\n",
       "                                                              egene_id  \\\n",
       "cluster_egene_credibleset                                                \n",
       "ENSG00000025708.13_ENSG00000177989.13_e_ENSG000...  ENSG00000025708.13   \n",
       "ENSG00000025708.13_ENSG00000177989.13_e_ENSG000...  ENSG00000025708.13   \n",
       "ENSG00000025708.13_ENSG00000177989.13_e_ENSG000...  ENSG00000177989.13   \n",
       "ENSG00000025708.13_ENSG00000177989.13_e_ENSG000...  ENSG00000177989.13   \n",
       "ENSG00000093000.18_ENSG00000100364.18_e_ENSG000...  ENSG00000093000.18   \n",
       "\n",
       "                                                                                          cs_id  \\\n",
       "cluster_egene_credibleset                                                                         \n",
       "ENSG00000025708.13_ENSG00000177989.13_e_ENSG000...  ENSG00000025708.13_ENSG00000177989.13_pc1_1   \n",
       "ENSG00000025708.13_ENSG00000177989.13_e_ENSG000...  ENSG00000025708.13_ENSG00000177989.13_pc1_2   \n",
       "ENSG00000025708.13_ENSG00000177989.13_e_ENSG000...  ENSG00000025708.13_ENSG00000177989.13_pc1_1   \n",
       "ENSG00000025708.13_ENSG00000177989.13_e_ENSG000...  ENSG00000025708.13_ENSG00000177989.13_pc1_2   \n",
       "ENSG00000093000.18_ENSG00000100364.18_e_ENSG000...  ENSG00000093000.18_ENSG00000100364.18_pc1_1   \n",
       "\n",
       "                                                           lead_variant_id  \\\n",
       "cluster_egene_credibleset                                                    \n",
       "ENSG00000025708.13_ENSG00000177989.13_e_ENSG000...  chr22_50532837_T_C_b38   \n",
       "ENSG00000025708.13_ENSG00000177989.13_e_ENSG000...  chr22_50519627_C_T_b38   \n",
       "ENSG00000025708.13_ENSG00000177989.13_e_ENSG000...  chr22_50532837_T_C_b38   \n",
       "ENSG00000025708.13_ENSG00000177989.13_e_ENSG000...  chr22_50519627_C_T_b38   \n",
       "ENSG00000093000.18_ENSG00000100364.18_e_ENSG000...  chr22_45105543_G_A_b38   \n",
       "\n",
       "                                                    num_vars  \n",
       "cluster_egene_credibleset                                     \n",
       "ENSG00000025708.13_ENSG00000177989.13_e_ENSG000...         1  \n",
       "ENSG00000025708.13_ENSG00000177989.13_e_ENSG000...        11  \n",
       "ENSG00000025708.13_ENSG00000177989.13_e_ENSG000...         1  \n",
       "ENSG00000025708.13_ENSG00000177989.13_e_ENSG000...        11  \n",
       "ENSG00000093000.18_ENSG00000100364.18_e_ENSG000...         1  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pcphenotype_evariance.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "now I do something similar for the pc_nominals, basically just adding to the susie df the pc var explained"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>start_distance</th>\n",
       "      <th>end_distance</th>\n",
       "      <th>af</th>\n",
       "      <th>ma_samples</th>\n",
       "      <th>ma_count</th>\n",
       "      <th>pval_nominal</th>\n",
       "      <th>slope</th>\n",
       "      <th>slope_se</th>\n",
       "      <th>...</th>\n",
       "      <th>var_cluster</th>\n",
       "      <th>var_pcgene_cluster</th>\n",
       "      <th>cs_id</th>\n",
       "      <th>pip</th>\n",
       "      <th>pc_num</th>\n",
       "      <th>lead_variant_id</th>\n",
       "      <th>num_vars</th>\n",
       "      <th>variance</th>\n",
       "      <th>variance_weighted</th>\n",
       "      <th>cluster_pc_credibleset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000100033.16_ENSG00000183628.12_pc1</td>\n",
       "      <td>chr22_18933556_A_G_b38</td>\n",
       "      <td>27528</td>\n",
       "      <td>-2854</td>\n",
       "      <td>0.111801</td>\n",
       "      <td>104</td>\n",
       "      <td>108</td>\n",
       "      <td>1.005330e-13</td>\n",
       "      <td>0.040882</td>\n",
       "      <td>0.005308</td>\n",
       "      <td>...</td>\n",
       "      <td>chr22_18933556_A_G_b38_ENSG00000100033.16_ENSG...</td>\n",
       "      <td>chr22_18933556_A_G_b38_ENSG00000100033.16_ENSG...</td>\n",
       "      <td>ENSG00000100033.16_ENSG00000183628.12_pc1_1</td>\n",
       "      <td>0.939348</td>\n",
       "      <td>1</td>\n",
       "      <td>chr22_18933556_A_G_b38</td>\n",
       "      <td>2</td>\n",
       "      <td>0.167132</td>\n",
       "      <td>0.156995</td>\n",
       "      <td>ENSG00000100033.16_ENSG00000183628.12_pc1_cs_p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000100033.16_ENSG00000183628.12_pc2</td>\n",
       "      <td>chr22_18933556_A_G_b38</td>\n",
       "      <td>27528</td>\n",
       "      <td>-2854</td>\n",
       "      <td>0.111801</td>\n",
       "      <td>104</td>\n",
       "      <td>108</td>\n",
       "      <td>6.405509e-06</td>\n",
       "      <td>-0.025318</td>\n",
       "      <td>0.005538</td>\n",
       "      <td>...</td>\n",
       "      <td>chr22_18933556_A_G_b38_ENSG00000100033.16_ENSG...</td>\n",
       "      <td>chr22_18933556_A_G_b38_ENSG00000100033.16_ENSG...</td>\n",
       "      <td>ENSG00000100033.16_ENSG00000183628.12_pc1_1</td>\n",
       "      <td>0.939348</td>\n",
       "      <td>1</td>\n",
       "      <td>chr22_18933556_A_G_b38</td>\n",
       "      <td>2</td>\n",
       "      <td>0.064103</td>\n",
       "      <td>0.060215</td>\n",
       "      <td>ENSG00000100033.16_ENSG00000183628.12_pc2_cs_p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000100033.16_ENSG00000183628.12_pc1</td>\n",
       "      <td>chr22_18940813_T_C_b38</td>\n",
       "      <td>34785</td>\n",
       "      <td>4403</td>\n",
       "      <td>0.110766</td>\n",
       "      <td>104</td>\n",
       "      <td>107</td>\n",
       "      <td>9.494485e-13</td>\n",
       "      <td>0.039920</td>\n",
       "      <td>0.005418</td>\n",
       "      <td>...</td>\n",
       "      <td>chr22_18940813_T_C_b38_ENSG00000100033.16_ENSG...</td>\n",
       "      <td>chr22_18940813_T_C_b38_ENSG00000100033.16_ENSG...</td>\n",
       "      <td>ENSG00000100033.16_ENSG00000183628.12_pc1_1</td>\n",
       "      <td>0.060652</td>\n",
       "      <td>1</td>\n",
       "      <td>chr22_18933556_A_G_b38</td>\n",
       "      <td>2</td>\n",
       "      <td>0.159358</td>\n",
       "      <td>0.009665</td>\n",
       "      <td>ENSG00000100033.16_ENSG00000183628.12_pc1_cs_p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000100033.16_ENSG00000183628.12_pc2</td>\n",
       "      <td>chr22_18940813_T_C_b38</td>\n",
       "      <td>34785</td>\n",
       "      <td>4403</td>\n",
       "      <td>0.110766</td>\n",
       "      <td>104</td>\n",
       "      <td>107</td>\n",
       "      <td>1.865919e-06</td>\n",
       "      <td>-0.027117</td>\n",
       "      <td>0.005606</td>\n",
       "      <td>...</td>\n",
       "      <td>chr22_18940813_T_C_b38_ENSG00000100033.16_ENSG...</td>\n",
       "      <td>chr22_18940813_T_C_b38_ENSG00000100033.16_ENSG...</td>\n",
       "      <td>ENSG00000100033.16_ENSG00000183628.12_pc1_1</td>\n",
       "      <td>0.060652</td>\n",
       "      <td>1</td>\n",
       "      <td>chr22_18933556_A_G_b38</td>\n",
       "      <td>2</td>\n",
       "      <td>0.073531</td>\n",
       "      <td>0.004460</td>\n",
       "      <td>ENSG00000100033.16_ENSG00000183628.12_pc2_cs_p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000099904.15_ENSG00000234409.6_pc1</td>\n",
       "      <td>chr22_20153539_C_A_b38</td>\n",
       "      <td>24083</td>\n",
       "      <td>2474</td>\n",
       "      <td>0.319876</td>\n",
       "      <td>261</td>\n",
       "      <td>309</td>\n",
       "      <td>1.132345e-48</td>\n",
       "      <td>0.048625</td>\n",
       "      <td>0.002893</td>\n",
       "      <td>...</td>\n",
       "      <td>chr22_20153539_C_A_b38_ENSG00000099904.15_ENSG...</td>\n",
       "      <td>chr22_20153539_C_A_b38_ENSG00000099904.15_ENSG...</td>\n",
       "      <td>ENSG00000099904.15_ENSG00000234409.6_pc1_1</td>\n",
       "      <td>0.142057</td>\n",
       "      <td>1</td>\n",
       "      <td>chr22_20154469_C_G_b38</td>\n",
       "      <td>3</td>\n",
       "      <td>0.236434</td>\n",
       "      <td>0.033587</td>\n",
       "      <td>ENSG00000099904.15_ENSG00000234409.6_pc1_cs_pc1_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_pc2</td>\n",
       "      <td>chr22_50535371_C_G_b38</td>\n",
       "      <td>5315</td>\n",
       "      <td>2791</td>\n",
       "      <td>0.021739</td>\n",
       "      <td>20</td>\n",
       "      <td>21</td>\n",
       "      <td>5.438037e-01</td>\n",
       "      <td>0.007987</td>\n",
       "      <td>0.013145</td>\n",
       "      <td>...</td>\n",
       "      <td>chr22_50535371_C_G_b38_ENSG00000025708.13_ENSG...</td>\n",
       "      <td>chr22_50535371_C_G_b38_ENSG00000025708.13_ENSG...</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_pc1_2</td>\n",
       "      <td>0.101850</td>\n",
       "      <td>1</td>\n",
       "      <td>chr22_50519627_C_T_b38</td>\n",
       "      <td>11</td>\n",
       "      <td>0.006379</td>\n",
       "      <td>0.000650</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_pc2_cs_p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_pc1</td>\n",
       "      <td>chr22_50540263_G_A_b38</td>\n",
       "      <td>10207</td>\n",
       "      <td>7683</td>\n",
       "      <td>0.022774</td>\n",
       "      <td>20</td>\n",
       "      <td>22</td>\n",
       "      <td>5.443152e-06</td>\n",
       "      <td>-0.055229</td>\n",
       "      <td>0.011987</td>\n",
       "      <td>...</td>\n",
       "      <td>chr22_50540263_G_A_b38_ENSG00000025708.13_ENSG...</td>\n",
       "      <td>chr22_50540263_G_A_b38_ENSG00000025708.13_ENSG...</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_pc1_2</td>\n",
       "      <td>0.088870</td>\n",
       "      <td>1</td>\n",
       "      <td>chr22_50519627_C_T_b38</td>\n",
       "      <td>11</td>\n",
       "      <td>0.305021</td>\n",
       "      <td>0.027107</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_pc1_cs_p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201</th>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_pc2</td>\n",
       "      <td>chr22_50540263_G_A_b38</td>\n",
       "      <td>10207</td>\n",
       "      <td>7683</td>\n",
       "      <td>0.022774</td>\n",
       "      <td>20</td>\n",
       "      <td>22</td>\n",
       "      <td>3.925436e-01</td>\n",
       "      <td>0.010511</td>\n",
       "      <td>0.012281</td>\n",
       "      <td>...</td>\n",
       "      <td>chr22_50540263_G_A_b38_ENSG00000025708.13_ENSG...</td>\n",
       "      <td>chr22_50540263_G_A_b38_ENSG00000025708.13_ENSG...</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_pc1_2</td>\n",
       "      <td>0.088870</td>\n",
       "      <td>1</td>\n",
       "      <td>chr22_50519627_C_T_b38</td>\n",
       "      <td>11</td>\n",
       "      <td>0.011048</td>\n",
       "      <td>0.000982</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_pc2_cs_p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>202</th>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_pc1</td>\n",
       "      <td>chr22_50541096_C_A_b38</td>\n",
       "      <td>11040</td>\n",
       "      <td>8516</td>\n",
       "      <td>0.021739</td>\n",
       "      <td>20</td>\n",
       "      <td>21</td>\n",
       "      <td>6.125603e-06</td>\n",
       "      <td>-0.058775</td>\n",
       "      <td>0.012829</td>\n",
       "      <td>...</td>\n",
       "      <td>chr22_50541096_C_A_b38_ENSG00000025708.13_ENSG...</td>\n",
       "      <td>chr22_50541096_C_A_b38_ENSG00000025708.13_ENSG...</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_pc1_2</td>\n",
       "      <td>0.101850</td>\n",
       "      <td>1</td>\n",
       "      <td>chr22_50519627_C_T_b38</td>\n",
       "      <td>11</td>\n",
       "      <td>0.345452</td>\n",
       "      <td>0.035184</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_pc1_cs_p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_pc2</td>\n",
       "      <td>chr22_50541096_C_A_b38</td>\n",
       "      <td>11040</td>\n",
       "      <td>8516</td>\n",
       "      <td>0.021739</td>\n",
       "      <td>20</td>\n",
       "      <td>21</td>\n",
       "      <td>5.438037e-01</td>\n",
       "      <td>0.007987</td>\n",
       "      <td>0.013145</td>\n",
       "      <td>...</td>\n",
       "      <td>chr22_50541096_C_A_b38_ENSG00000025708.13_ENSG...</td>\n",
       "      <td>chr22_50541096_C_A_b38_ENSG00000025708.13_ENSG...</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_pc1_2</td>\n",
       "      <td>0.101850</td>\n",
       "      <td>1</td>\n",
       "      <td>chr22_50519627_C_T_b38</td>\n",
       "      <td>11</td>\n",
       "      <td>0.006379</td>\n",
       "      <td>0.000650</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_pc2_cs_p...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>204 rows Ã— 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  phenotype_id              variant_id  \\\n",
       "0    ENSG00000100033.16_ENSG00000183628.12_pc1  chr22_18933556_A_G_b38   \n",
       "1    ENSG00000100033.16_ENSG00000183628.12_pc2  chr22_18933556_A_G_b38   \n",
       "2    ENSG00000100033.16_ENSG00000183628.12_pc1  chr22_18940813_T_C_b38   \n",
       "3    ENSG00000100033.16_ENSG00000183628.12_pc2  chr22_18940813_T_C_b38   \n",
       "4     ENSG00000099904.15_ENSG00000234409.6_pc1  chr22_20153539_C_A_b38   \n",
       "..                                         ...                     ...   \n",
       "199  ENSG00000025708.13_ENSG00000177989.13_pc2  chr22_50535371_C_G_b38   \n",
       "200  ENSG00000025708.13_ENSG00000177989.13_pc1  chr22_50540263_G_A_b38   \n",
       "201  ENSG00000025708.13_ENSG00000177989.13_pc2  chr22_50540263_G_A_b38   \n",
       "202  ENSG00000025708.13_ENSG00000177989.13_pc1  chr22_50541096_C_A_b38   \n",
       "203  ENSG00000025708.13_ENSG00000177989.13_pc2  chr22_50541096_C_A_b38   \n",
       "\n",
       "     start_distance  end_distance        af  ma_samples  ma_count  \\\n",
       "0             27528         -2854  0.111801         104       108   \n",
       "1             27528         -2854  0.111801         104       108   \n",
       "2             34785          4403  0.110766         104       107   \n",
       "3             34785          4403  0.110766         104       107   \n",
       "4             24083          2474  0.319876         261       309   \n",
       "..              ...           ...       ...         ...       ...   \n",
       "199            5315          2791  0.021739          20        21   \n",
       "200           10207          7683  0.022774          20        22   \n",
       "201           10207          7683  0.022774          20        22   \n",
       "202           11040          8516  0.021739          20        21   \n",
       "203           11040          8516  0.021739          20        21   \n",
       "\n",
       "     pval_nominal     slope  slope_se  ...  \\\n",
       "0    1.005330e-13  0.040882  0.005308  ...   \n",
       "1    6.405509e-06 -0.025318  0.005538  ...   \n",
       "2    9.494485e-13  0.039920  0.005418  ...   \n",
       "3    1.865919e-06 -0.027117  0.005606  ...   \n",
       "4    1.132345e-48  0.048625  0.002893  ...   \n",
       "..            ...       ...       ...  ...   \n",
       "199  5.438037e-01  0.007987  0.013145  ...   \n",
       "200  5.443152e-06 -0.055229  0.011987  ...   \n",
       "201  3.925436e-01  0.010511  0.012281  ...   \n",
       "202  6.125603e-06 -0.058775  0.012829  ...   \n",
       "203  5.438037e-01  0.007987  0.013145  ...   \n",
       "\n",
       "                                           var_cluster  \\\n",
       "0    chr22_18933556_A_G_b38_ENSG00000100033.16_ENSG...   \n",
       "1    chr22_18933556_A_G_b38_ENSG00000100033.16_ENSG...   \n",
       "2    chr22_18940813_T_C_b38_ENSG00000100033.16_ENSG...   \n",
       "3    chr22_18940813_T_C_b38_ENSG00000100033.16_ENSG...   \n",
       "4    chr22_20153539_C_A_b38_ENSG00000099904.15_ENSG...   \n",
       "..                                                 ...   \n",
       "199  chr22_50535371_C_G_b38_ENSG00000025708.13_ENSG...   \n",
       "200  chr22_50540263_G_A_b38_ENSG00000025708.13_ENSG...   \n",
       "201  chr22_50540263_G_A_b38_ENSG00000025708.13_ENSG...   \n",
       "202  chr22_50541096_C_A_b38_ENSG00000025708.13_ENSG...   \n",
       "203  chr22_50541096_C_A_b38_ENSG00000025708.13_ENSG...   \n",
       "\n",
       "                                    var_pcgene_cluster  \\\n",
       "0    chr22_18933556_A_G_b38_ENSG00000100033.16_ENSG...   \n",
       "1    chr22_18933556_A_G_b38_ENSG00000100033.16_ENSG...   \n",
       "2    chr22_18940813_T_C_b38_ENSG00000100033.16_ENSG...   \n",
       "3    chr22_18940813_T_C_b38_ENSG00000100033.16_ENSG...   \n",
       "4    chr22_20153539_C_A_b38_ENSG00000099904.15_ENSG...   \n",
       "..                                                 ...   \n",
       "199  chr22_50535371_C_G_b38_ENSG00000025708.13_ENSG...   \n",
       "200  chr22_50540263_G_A_b38_ENSG00000025708.13_ENSG...   \n",
       "201  chr22_50540263_G_A_b38_ENSG00000025708.13_ENSG...   \n",
       "202  chr22_50541096_C_A_b38_ENSG00000025708.13_ENSG...   \n",
       "203  chr22_50541096_C_A_b38_ENSG00000025708.13_ENSG...   \n",
       "\n",
       "                                           cs_id       pip  pc_num  \\\n",
       "0    ENSG00000100033.16_ENSG00000183628.12_pc1_1  0.939348       1   \n",
       "1    ENSG00000100033.16_ENSG00000183628.12_pc1_1  0.939348       1   \n",
       "2    ENSG00000100033.16_ENSG00000183628.12_pc1_1  0.060652       1   \n",
       "3    ENSG00000100033.16_ENSG00000183628.12_pc1_1  0.060652       1   \n",
       "4     ENSG00000099904.15_ENSG00000234409.6_pc1_1  0.142057       1   \n",
       "..                                           ...       ...     ...   \n",
       "199  ENSG00000025708.13_ENSG00000177989.13_pc1_2  0.101850       1   \n",
       "200  ENSG00000025708.13_ENSG00000177989.13_pc1_2  0.088870       1   \n",
       "201  ENSG00000025708.13_ENSG00000177989.13_pc1_2  0.088870       1   \n",
       "202  ENSG00000025708.13_ENSG00000177989.13_pc1_2  0.101850       1   \n",
       "203  ENSG00000025708.13_ENSG00000177989.13_pc1_2  0.101850       1   \n",
       "\n",
       "            lead_variant_id num_vars  variance  variance_weighted  \\\n",
       "0    chr22_18933556_A_G_b38        2  0.167132           0.156995   \n",
       "1    chr22_18933556_A_G_b38        2  0.064103           0.060215   \n",
       "2    chr22_18933556_A_G_b38        2  0.159358           0.009665   \n",
       "3    chr22_18933556_A_G_b38        2  0.073531           0.004460   \n",
       "4    chr22_20154469_C_G_b38        3  0.236434           0.033587   \n",
       "..                      ...      ...       ...                ...   \n",
       "199  chr22_50519627_C_T_b38       11  0.006379           0.000650   \n",
       "200  chr22_50519627_C_T_b38       11  0.305021           0.027107   \n",
       "201  chr22_50519627_C_T_b38       11  0.011048           0.000982   \n",
       "202  chr22_50519627_C_T_b38       11  0.345452           0.035184   \n",
       "203  chr22_50519627_C_T_b38       11  0.006379           0.000650   \n",
       "\n",
       "                                cluster_pc_credibleset  \n",
       "0    ENSG00000100033.16_ENSG00000183628.12_pc1_cs_p...  \n",
       "1    ENSG00000100033.16_ENSG00000183628.12_pc2_cs_p...  \n",
       "2    ENSG00000100033.16_ENSG00000183628.12_pc1_cs_p...  \n",
       "3    ENSG00000100033.16_ENSG00000183628.12_pc2_cs_p...  \n",
       "4    ENSG00000099904.15_ENSG00000234409.6_pc1_cs_pc1_1  \n",
       "..                                                 ...  \n",
       "199  ENSG00000025708.13_ENSG00000177989.13_pc2_cs_p...  \n",
       "200  ENSG00000025708.13_ENSG00000177989.13_pc1_cs_p...  \n",
       "201  ENSG00000025708.13_ENSG00000177989.13_pc2_cs_p...  \n",
       "202  ENSG00000025708.13_ENSG00000177989.13_pc1_cs_p...  \n",
       "203  ENSG00000025708.13_ENSG00000177989.13_pc2_cs_p...  \n",
       "\n",
       "[204 rows x 21 columns]"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pc_nominal_pcsusie_subset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/local/scratch/klawren/slrmtmp.44431414/ipykernel_1284/787166930.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  pc_nominal_pcsusie_subset['var_pcgene_cluster'] = pc_nominal_pcsusie_subset['variant_id'] + '_' + pc_nominal_pcsusie_subset['phenotype_id']\n"
     ]
    }
   ],
   "source": [
    "# get a subset of variants that ended up in credible sets\n",
    "pc_nominal_pcsusie_subset = pc_nominal[pc_nominal['var_cluster'].isin(pc_susie['var_cluster'])]\n",
    "\n",
    "# I want to get a pip weighted variance for each egene\n",
    "pc_nominal_pcsusie_subset['var_pcgene_cluster'] = pc_nominal_pcsusie_subset['variant_id'] + '_' + pc_nominal_pcsusie_subset['phenotype_id']\n",
    "\n",
    "# merge in the data from the susie finemapping\n",
    "pc_nominal_pcsusie_subset = pd.merge(pc_nominal_pcsusie_subset, pc_susie[['cs_id', 'var_cluster', 'pip', 'pc_num', 'lead_variant_id', 'num_vars']], on='var_cluster')\n",
    "\n",
    "# pip weighted variance\n",
    "pc_nominal_pcsusie_subset['variance'] = pc_nominal_pcsusie_subset['slope'].apply(np.square) * 100\n",
    "pc_nominal_pcsusie_subset['variance_weighted'] = pc_nominal_pcsusie_subset['variance'] * e_nominal_pcsusie_subset['pip'] \n",
    "\n",
    "# we should group based on unique sets of variants belonging to a certain cluster, and credible set\n",
    "# because it only makes sese to look at vars for the pc that the credible set is for, we drop the irrelevant pc-pc pairings\n",
    "pc_nominal_pcsusie_subset['cluster_pc_credibleset'] = pc_nominal_pcsusie_subset['phenotype_id'] + '_cs' + '_pc' + pc_nominal_pcsusie_subset['cs_id'].str.split('_pc').str[-1]\n",
    "\n",
    "# group by phenotype (egene and cluster)\n",
    "pcphenotype_pcvariance = pc_nominal_pcsusie_subset.groupby('cluster_pc_credibleset').agg({'variance_weighted':sum, \n",
    "                                                        'cluster_id':'first', \n",
    "                                                        'variant_id':list,\n",
    "                                                        'cs_id':'first', \n",
    "                                                        'lead_variant_id':'first',\n",
    "                                                        'num_vars':'first',\n",
    "                                                        'pc_num':list})\n",
    "\n",
    "pcphenotype_pcvariance = pcphenotype_pcvariance.rename(columns={'variant_id':'variant_list'})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>start_distance</th>\n",
       "      <th>end_distance</th>\n",
       "      <th>af</th>\n",
       "      <th>ma_samples</th>\n",
       "      <th>ma_count</th>\n",
       "      <th>pval_nominal</th>\n",
       "      <th>slope</th>\n",
       "      <th>slope_se</th>\n",
       "      <th>...</th>\n",
       "      <th>var_cluster</th>\n",
       "      <th>var_pcgene_cluster</th>\n",
       "      <th>cs_id</th>\n",
       "      <th>pip</th>\n",
       "      <th>pc_num</th>\n",
       "      <th>lead_variant_id</th>\n",
       "      <th>num_vars</th>\n",
       "      <th>variance</th>\n",
       "      <th>variance_weighted</th>\n",
       "      <th>cluster_pc_credibleset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000100033.16_ENSG00000183628.12_pc1</td>\n",
       "      <td>chr22_18933556_A_G_b38</td>\n",
       "      <td>27528</td>\n",
       "      <td>-2854</td>\n",
       "      <td>0.111801</td>\n",
       "      <td>104</td>\n",
       "      <td>108</td>\n",
       "      <td>1.005330e-13</td>\n",
       "      <td>0.040882</td>\n",
       "      <td>0.005308</td>\n",
       "      <td>...</td>\n",
       "      <td>chr22_18933556_A_G_b38_ENSG00000100033.16_ENSG...</td>\n",
       "      <td>chr22_18933556_A_G_b38_ENSG00000100033.16_ENSG...</td>\n",
       "      <td>ENSG00000100033.16_ENSG00000183628.12_pc1_1</td>\n",
       "      <td>0.939348</td>\n",
       "      <td>1</td>\n",
       "      <td>chr22_18933556_A_G_b38</td>\n",
       "      <td>2</td>\n",
       "      <td>0.167132</td>\n",
       "      <td>0.156995</td>\n",
       "      <td>ENSG00000100033.16_ENSG00000183628.12_pc1_cs_p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000100033.16_ENSG00000183628.12_pc2</td>\n",
       "      <td>chr22_18933556_A_G_b38</td>\n",
       "      <td>27528</td>\n",
       "      <td>-2854</td>\n",
       "      <td>0.111801</td>\n",
       "      <td>104</td>\n",
       "      <td>108</td>\n",
       "      <td>6.405509e-06</td>\n",
       "      <td>-0.025318</td>\n",
       "      <td>0.005538</td>\n",
       "      <td>...</td>\n",
       "      <td>chr22_18933556_A_G_b38_ENSG00000100033.16_ENSG...</td>\n",
       "      <td>chr22_18933556_A_G_b38_ENSG00000100033.16_ENSG...</td>\n",
       "      <td>ENSG00000100033.16_ENSG00000183628.12_pc1_1</td>\n",
       "      <td>0.939348</td>\n",
       "      <td>1</td>\n",
       "      <td>chr22_18933556_A_G_b38</td>\n",
       "      <td>2</td>\n",
       "      <td>0.064103</td>\n",
       "      <td>0.060215</td>\n",
       "      <td>ENSG00000100033.16_ENSG00000183628.12_pc2_cs_p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000100033.16_ENSG00000183628.12_pc1</td>\n",
       "      <td>chr22_18940813_T_C_b38</td>\n",
       "      <td>34785</td>\n",
       "      <td>4403</td>\n",
       "      <td>0.110766</td>\n",
       "      <td>104</td>\n",
       "      <td>107</td>\n",
       "      <td>9.494485e-13</td>\n",
       "      <td>0.039920</td>\n",
       "      <td>0.005418</td>\n",
       "      <td>...</td>\n",
       "      <td>chr22_18940813_T_C_b38_ENSG00000100033.16_ENSG...</td>\n",
       "      <td>chr22_18940813_T_C_b38_ENSG00000100033.16_ENSG...</td>\n",
       "      <td>ENSG00000100033.16_ENSG00000183628.12_pc1_1</td>\n",
       "      <td>0.060652</td>\n",
       "      <td>1</td>\n",
       "      <td>chr22_18933556_A_G_b38</td>\n",
       "      <td>2</td>\n",
       "      <td>0.159358</td>\n",
       "      <td>0.009665</td>\n",
       "      <td>ENSG00000100033.16_ENSG00000183628.12_pc1_cs_p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000100033.16_ENSG00000183628.12_pc2</td>\n",
       "      <td>chr22_18940813_T_C_b38</td>\n",
       "      <td>34785</td>\n",
       "      <td>4403</td>\n",
       "      <td>0.110766</td>\n",
       "      <td>104</td>\n",
       "      <td>107</td>\n",
       "      <td>1.865919e-06</td>\n",
       "      <td>-0.027117</td>\n",
       "      <td>0.005606</td>\n",
       "      <td>...</td>\n",
       "      <td>chr22_18940813_T_C_b38_ENSG00000100033.16_ENSG...</td>\n",
       "      <td>chr22_18940813_T_C_b38_ENSG00000100033.16_ENSG...</td>\n",
       "      <td>ENSG00000100033.16_ENSG00000183628.12_pc1_1</td>\n",
       "      <td>0.060652</td>\n",
       "      <td>1</td>\n",
       "      <td>chr22_18933556_A_G_b38</td>\n",
       "      <td>2</td>\n",
       "      <td>0.073531</td>\n",
       "      <td>0.004460</td>\n",
       "      <td>ENSG00000100033.16_ENSG00000183628.12_pc2_cs_p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000099904.15_ENSG00000234409.6_pc1</td>\n",
       "      <td>chr22_20153539_C_A_b38</td>\n",
       "      <td>24083</td>\n",
       "      <td>2474</td>\n",
       "      <td>0.319876</td>\n",
       "      <td>261</td>\n",
       "      <td>309</td>\n",
       "      <td>1.132345e-48</td>\n",
       "      <td>0.048625</td>\n",
       "      <td>0.002893</td>\n",
       "      <td>...</td>\n",
       "      <td>chr22_20153539_C_A_b38_ENSG00000099904.15_ENSG...</td>\n",
       "      <td>chr22_20153539_C_A_b38_ENSG00000099904.15_ENSG...</td>\n",
       "      <td>ENSG00000099904.15_ENSG00000234409.6_pc1_1</td>\n",
       "      <td>0.142057</td>\n",
       "      <td>1</td>\n",
       "      <td>chr22_20154469_C_G_b38</td>\n",
       "      <td>3</td>\n",
       "      <td>0.236434</td>\n",
       "      <td>0.033587</td>\n",
       "      <td>ENSG00000099904.15_ENSG00000234409.6_pc1_cs_pc1_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_pc2</td>\n",
       "      <td>chr22_50535371_C_G_b38</td>\n",
       "      <td>5315</td>\n",
       "      <td>2791</td>\n",
       "      <td>0.021739</td>\n",
       "      <td>20</td>\n",
       "      <td>21</td>\n",
       "      <td>5.438037e-01</td>\n",
       "      <td>0.007987</td>\n",
       "      <td>0.013145</td>\n",
       "      <td>...</td>\n",
       "      <td>chr22_50535371_C_G_b38_ENSG00000025708.13_ENSG...</td>\n",
       "      <td>chr22_50535371_C_G_b38_ENSG00000025708.13_ENSG...</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_pc1_2</td>\n",
       "      <td>0.101850</td>\n",
       "      <td>1</td>\n",
       "      <td>chr22_50519627_C_T_b38</td>\n",
       "      <td>11</td>\n",
       "      <td>0.006379</td>\n",
       "      <td>0.000650</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_pc2_cs_p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_pc1</td>\n",
       "      <td>chr22_50540263_G_A_b38</td>\n",
       "      <td>10207</td>\n",
       "      <td>7683</td>\n",
       "      <td>0.022774</td>\n",
       "      <td>20</td>\n",
       "      <td>22</td>\n",
       "      <td>5.443152e-06</td>\n",
       "      <td>-0.055229</td>\n",
       "      <td>0.011987</td>\n",
       "      <td>...</td>\n",
       "      <td>chr22_50540263_G_A_b38_ENSG00000025708.13_ENSG...</td>\n",
       "      <td>chr22_50540263_G_A_b38_ENSG00000025708.13_ENSG...</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_pc1_2</td>\n",
       "      <td>0.088870</td>\n",
       "      <td>1</td>\n",
       "      <td>chr22_50519627_C_T_b38</td>\n",
       "      <td>11</td>\n",
       "      <td>0.305021</td>\n",
       "      <td>0.027107</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_pc1_cs_p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201</th>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_pc2</td>\n",
       "      <td>chr22_50540263_G_A_b38</td>\n",
       "      <td>10207</td>\n",
       "      <td>7683</td>\n",
       "      <td>0.022774</td>\n",
       "      <td>20</td>\n",
       "      <td>22</td>\n",
       "      <td>3.925436e-01</td>\n",
       "      <td>0.010511</td>\n",
       "      <td>0.012281</td>\n",
       "      <td>...</td>\n",
       "      <td>chr22_50540263_G_A_b38_ENSG00000025708.13_ENSG...</td>\n",
       "      <td>chr22_50540263_G_A_b38_ENSG00000025708.13_ENSG...</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_pc1_2</td>\n",
       "      <td>0.088870</td>\n",
       "      <td>1</td>\n",
       "      <td>chr22_50519627_C_T_b38</td>\n",
       "      <td>11</td>\n",
       "      <td>0.011048</td>\n",
       "      <td>0.000982</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_pc2_cs_p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>202</th>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_pc1</td>\n",
       "      <td>chr22_50541096_C_A_b38</td>\n",
       "      <td>11040</td>\n",
       "      <td>8516</td>\n",
       "      <td>0.021739</td>\n",
       "      <td>20</td>\n",
       "      <td>21</td>\n",
       "      <td>6.125603e-06</td>\n",
       "      <td>-0.058775</td>\n",
       "      <td>0.012829</td>\n",
       "      <td>...</td>\n",
       "      <td>chr22_50541096_C_A_b38_ENSG00000025708.13_ENSG...</td>\n",
       "      <td>chr22_50541096_C_A_b38_ENSG00000025708.13_ENSG...</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_pc1_2</td>\n",
       "      <td>0.101850</td>\n",
       "      <td>1</td>\n",
       "      <td>chr22_50519627_C_T_b38</td>\n",
       "      <td>11</td>\n",
       "      <td>0.345452</td>\n",
       "      <td>0.035184</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_pc1_cs_p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_pc2</td>\n",
       "      <td>chr22_50541096_C_A_b38</td>\n",
       "      <td>11040</td>\n",
       "      <td>8516</td>\n",
       "      <td>0.021739</td>\n",
       "      <td>20</td>\n",
       "      <td>21</td>\n",
       "      <td>5.438037e-01</td>\n",
       "      <td>0.007987</td>\n",
       "      <td>0.013145</td>\n",
       "      <td>...</td>\n",
       "      <td>chr22_50541096_C_A_b38_ENSG00000025708.13_ENSG...</td>\n",
       "      <td>chr22_50541096_C_A_b38_ENSG00000025708.13_ENSG...</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_pc1_2</td>\n",
       "      <td>0.101850</td>\n",
       "      <td>1</td>\n",
       "      <td>chr22_50519627_C_T_b38</td>\n",
       "      <td>11</td>\n",
       "      <td>0.006379</td>\n",
       "      <td>0.000650</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_pc2_cs_p...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>204 rows Ã— 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  phenotype_id              variant_id  \\\n",
       "0    ENSG00000100033.16_ENSG00000183628.12_pc1  chr22_18933556_A_G_b38   \n",
       "1    ENSG00000100033.16_ENSG00000183628.12_pc2  chr22_18933556_A_G_b38   \n",
       "2    ENSG00000100033.16_ENSG00000183628.12_pc1  chr22_18940813_T_C_b38   \n",
       "3    ENSG00000100033.16_ENSG00000183628.12_pc2  chr22_18940813_T_C_b38   \n",
       "4     ENSG00000099904.15_ENSG00000234409.6_pc1  chr22_20153539_C_A_b38   \n",
       "..                                         ...                     ...   \n",
       "199  ENSG00000025708.13_ENSG00000177989.13_pc2  chr22_50535371_C_G_b38   \n",
       "200  ENSG00000025708.13_ENSG00000177989.13_pc1  chr22_50540263_G_A_b38   \n",
       "201  ENSG00000025708.13_ENSG00000177989.13_pc2  chr22_50540263_G_A_b38   \n",
       "202  ENSG00000025708.13_ENSG00000177989.13_pc1  chr22_50541096_C_A_b38   \n",
       "203  ENSG00000025708.13_ENSG00000177989.13_pc2  chr22_50541096_C_A_b38   \n",
       "\n",
       "     start_distance  end_distance        af  ma_samples  ma_count  \\\n",
       "0             27528         -2854  0.111801         104       108   \n",
       "1             27528         -2854  0.111801         104       108   \n",
       "2             34785          4403  0.110766         104       107   \n",
       "3             34785          4403  0.110766         104       107   \n",
       "4             24083          2474  0.319876         261       309   \n",
       "..              ...           ...       ...         ...       ...   \n",
       "199            5315          2791  0.021739          20        21   \n",
       "200           10207          7683  0.022774          20        22   \n",
       "201           10207          7683  0.022774          20        22   \n",
       "202           11040          8516  0.021739          20        21   \n",
       "203           11040          8516  0.021739          20        21   \n",
       "\n",
       "     pval_nominal     slope  slope_se  ...  \\\n",
       "0    1.005330e-13  0.040882  0.005308  ...   \n",
       "1    6.405509e-06 -0.025318  0.005538  ...   \n",
       "2    9.494485e-13  0.039920  0.005418  ...   \n",
       "3    1.865919e-06 -0.027117  0.005606  ...   \n",
       "4    1.132345e-48  0.048625  0.002893  ...   \n",
       "..            ...       ...       ...  ...   \n",
       "199  5.438037e-01  0.007987  0.013145  ...   \n",
       "200  5.443152e-06 -0.055229  0.011987  ...   \n",
       "201  3.925436e-01  0.010511  0.012281  ...   \n",
       "202  6.125603e-06 -0.058775  0.012829  ...   \n",
       "203  5.438037e-01  0.007987  0.013145  ...   \n",
       "\n",
       "                                           var_cluster  \\\n",
       "0    chr22_18933556_A_G_b38_ENSG00000100033.16_ENSG...   \n",
       "1    chr22_18933556_A_G_b38_ENSG00000100033.16_ENSG...   \n",
       "2    chr22_18940813_T_C_b38_ENSG00000100033.16_ENSG...   \n",
       "3    chr22_18940813_T_C_b38_ENSG00000100033.16_ENSG...   \n",
       "4    chr22_20153539_C_A_b38_ENSG00000099904.15_ENSG...   \n",
       "..                                                 ...   \n",
       "199  chr22_50535371_C_G_b38_ENSG00000025708.13_ENSG...   \n",
       "200  chr22_50540263_G_A_b38_ENSG00000025708.13_ENSG...   \n",
       "201  chr22_50540263_G_A_b38_ENSG00000025708.13_ENSG...   \n",
       "202  chr22_50541096_C_A_b38_ENSG00000025708.13_ENSG...   \n",
       "203  chr22_50541096_C_A_b38_ENSG00000025708.13_ENSG...   \n",
       "\n",
       "                                    var_pcgene_cluster  \\\n",
       "0    chr22_18933556_A_G_b38_ENSG00000100033.16_ENSG...   \n",
       "1    chr22_18933556_A_G_b38_ENSG00000100033.16_ENSG...   \n",
       "2    chr22_18940813_T_C_b38_ENSG00000100033.16_ENSG...   \n",
       "3    chr22_18940813_T_C_b38_ENSG00000100033.16_ENSG...   \n",
       "4    chr22_20153539_C_A_b38_ENSG00000099904.15_ENSG...   \n",
       "..                                                 ...   \n",
       "199  chr22_50535371_C_G_b38_ENSG00000025708.13_ENSG...   \n",
       "200  chr22_50540263_G_A_b38_ENSG00000025708.13_ENSG...   \n",
       "201  chr22_50540263_G_A_b38_ENSG00000025708.13_ENSG...   \n",
       "202  chr22_50541096_C_A_b38_ENSG00000025708.13_ENSG...   \n",
       "203  chr22_50541096_C_A_b38_ENSG00000025708.13_ENSG...   \n",
       "\n",
       "                                           cs_id       pip  pc_num  \\\n",
       "0    ENSG00000100033.16_ENSG00000183628.12_pc1_1  0.939348       1   \n",
       "1    ENSG00000100033.16_ENSG00000183628.12_pc1_1  0.939348       1   \n",
       "2    ENSG00000100033.16_ENSG00000183628.12_pc1_1  0.060652       1   \n",
       "3    ENSG00000100033.16_ENSG00000183628.12_pc1_1  0.060652       1   \n",
       "4     ENSG00000099904.15_ENSG00000234409.6_pc1_1  0.142057       1   \n",
       "..                                           ...       ...     ...   \n",
       "199  ENSG00000025708.13_ENSG00000177989.13_pc1_2  0.101850       1   \n",
       "200  ENSG00000025708.13_ENSG00000177989.13_pc1_2  0.088870       1   \n",
       "201  ENSG00000025708.13_ENSG00000177989.13_pc1_2  0.088870       1   \n",
       "202  ENSG00000025708.13_ENSG00000177989.13_pc1_2  0.101850       1   \n",
       "203  ENSG00000025708.13_ENSG00000177989.13_pc1_2  0.101850       1   \n",
       "\n",
       "            lead_variant_id num_vars  variance  variance_weighted  \\\n",
       "0    chr22_18933556_A_G_b38        2  0.167132           0.156995   \n",
       "1    chr22_18933556_A_G_b38        2  0.064103           0.060215   \n",
       "2    chr22_18933556_A_G_b38        2  0.159358           0.009665   \n",
       "3    chr22_18933556_A_G_b38        2  0.073531           0.004460   \n",
       "4    chr22_20154469_C_G_b38        3  0.236434           0.033587   \n",
       "..                      ...      ...       ...                ...   \n",
       "199  chr22_50519627_C_T_b38       11  0.006379           0.000650   \n",
       "200  chr22_50519627_C_T_b38       11  0.305021           0.027107   \n",
       "201  chr22_50519627_C_T_b38       11  0.011048           0.000982   \n",
       "202  chr22_50519627_C_T_b38       11  0.345452           0.035184   \n",
       "203  chr22_50519627_C_T_b38       11  0.006379           0.000650   \n",
       "\n",
       "                                cluster_pc_credibleset  \n",
       "0    ENSG00000100033.16_ENSG00000183628.12_pc1_cs_p...  \n",
       "1    ENSG00000100033.16_ENSG00000183628.12_pc2_cs_p...  \n",
       "2    ENSG00000100033.16_ENSG00000183628.12_pc1_cs_p...  \n",
       "3    ENSG00000100033.16_ENSG00000183628.12_pc2_cs_p...  \n",
       "4    ENSG00000099904.15_ENSG00000234409.6_pc1_cs_pc1_1  \n",
       "..                                                 ...  \n",
       "199  ENSG00000025708.13_ENSG00000177989.13_pc2_cs_p...  \n",
       "200  ENSG00000025708.13_ENSG00000177989.13_pc1_cs_p...  \n",
       "201  ENSG00000025708.13_ENSG00000177989.13_pc2_cs_p...  \n",
       "202  ENSG00000025708.13_ENSG00000177989.13_pc1_cs_p...  \n",
       "203  ENSG00000025708.13_ENSG00000177989.13_pc2_cs_p...  \n",
       "\n",
       "[204 rows x 21 columns]"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pc_nominal_pcsusie_subset['pc_num_variant'] = pc_nominal_pcsusie_subset['pc_num']\n",
    "pc_nominal_pcsusie_subset['pc_num_cs'] = pc_nominal_pcsusie_subset['pc_num']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "cluster_pc_credibleset\n",
       "ENSG00000025708.13_ENSG00000177989.13_pc1_cs_pc1_1                                                                                      [1]\n",
       "ENSG00000025708.13_ENSG00000177989.13_pc1_cs_pc1_2                                                        [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]\n",
       "ENSG00000025708.13_ENSG00000177989.13_pc2_cs_pc1_1                                                                                      [1]\n",
       "ENSG00000025708.13_ENSG00000177989.13_pc2_cs_pc1_2                                                        [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]\n",
       "ENSG00000093000.18_ENSG00000100364.18_pc1_cs_pc1_1                                                                                      [1]\n",
       "ENSG00000093000.18_ENSG00000100364.18_pc2_cs_pc1_1                                                                                      [1]\n",
       "ENSG00000099904.15_ENSG00000234409.6_pc1_cs_pc1_1                                                                                 [1, 1, 1]\n",
       "ENSG00000099904.15_ENSG00000234409.6_pc2_cs_pc1_1                                                                                 [1, 1, 1]\n",
       "ENSG00000099974.7_ENSG00000099977.13_ENSG00000133433.10_ENSG00000240972.1_pc1_cs_pc1_1                                               [1, 1]\n",
       "ENSG00000099974.7_ENSG00000099977.13_ENSG00000133433.10_ENSG00000240972.1_pc2_cs_pc1_1                                               [1, 1]\n",
       "ENSG00000099974.7_ENSG00000099977.13_ENSG00000133433.10_ENSG00000240972.1_pc3_cs_pc1_1                                               [1, 1]\n",
       "ENSG00000099974.7_ENSG00000099977.13_ENSG00000133433.10_ENSG00000240972.1_pc4_cs_pc1_1                                               [1, 1]\n",
       "ENSG00000099994.10_ENSG00000099998.17_pc1_cs_pc1_1                                                        [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]\n",
       "ENSG00000099994.10_ENSG00000099998.17_pc2_cs_pc1_1                                                        [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]\n",
       "ENSG00000100033.16_ENSG00000183628.12_pc1_cs_pc1_1                                                                                   [1, 1]\n",
       "ENSG00000100033.16_ENSG00000183628.12_pc2_cs_pc1_1                                                                                   [1, 1]\n",
       "ENSG00000100099.20_ENSG00000128203.6_pc1_cs_pc1_1                                                                                       [1]\n",
       "ENSG00000100099.20_ENSG00000128203.6_pc2_cs_pc1_1                                                                                       [1]\n",
       "ENSG00000100225.17_ENSG00000100234.11_pc1_cs_pc1_1                                                                                      [1]\n",
       "ENSG00000100225.17_ENSG00000100234.11_pc2_cs_pc1_1                                                                                      [1]\n",
       "ENSG00000100344.10_ENSG00000130540.13_pc1_cs_pc1_1                                                                             [1, 1, 1, 1]\n",
       "ENSG00000100344.10_ENSG00000130540.13_pc2_cs_pc1_1                                                                             [1, 1, 1, 1]\n",
       "ENSG00000128309.16_ENSG00000128311.13_pc1_cs_pc1_1                                                                                [1, 1, 1]\n",
       "ENSG00000128309.16_ENSG00000128311.13_pc1_cs_pc1_2                                                                                   [1, 1]\n",
       "ENSG00000128309.16_ENSG00000128311.13_pc2_cs_pc1_1                                                                                [1, 1, 1]\n",
       "ENSG00000128309.16_ENSG00000128311.13_pc2_cs_pc1_2                                                                                   [1, 1]\n",
       "ENSG00000128394.16_ENSG00000239713.8_pc1_cs_pc1_1                                                                                       [1]\n",
       "ENSG00000128394.16_ENSG00000239713.8_pc2_cs_pc1_1                                                                                       [1]\n",
       "ENSG00000177096.8_ENSG00000183172.8_pc1_cs_pc1_1                                                    [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]\n",
       "ENSG00000177096.8_ENSG00000183172.8_pc2_cs_pc1_1                                                    [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]\n",
       "ENSG00000183066.14_ENSG00000198951.11_pc1_cs_pc1_1                                        [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...\n",
       "ENSG00000183066.14_ENSG00000198951.11_pc1_cs_pc1_2                                                              [1, 1, 1, 1, 1, 1, 1, 1, 1]\n",
       "ENSG00000183066.14_ENSG00000198951.11_pc2_cs_pc1_1                                        [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...\n",
       "ENSG00000183066.14_ENSG00000198951.11_pc2_cs_pc1_2                                                              [1, 1, 1, 1, 1, 1, 1, 1, 1]\n",
       "ENSG00000243811.8_ENSG00000244509.3_pc1_cs_pc1_1                                                                                  [1, 1, 1]\n",
       "ENSG00000243811.8_ENSG00000244509.3_pc1_cs_pc1_2                                                             [1, 1, 1, 1, 1, 1, 1, 1, 1, 1]\n",
       "ENSG00000243811.8_ENSG00000244509.3_pc2_cs_pc1_1                                                                                  [1, 1, 1]\n",
       "ENSG00000243811.8_ENSG00000244509.3_pc2_cs_pc1_2                                                             [1, 1, 1, 1, 1, 1, 1, 1, 1, 1]\n",
       "Name: pc_num, dtype: object"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pcphenotype_pcvariance['pc_num']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'cluster_pc_credibleset'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m~/micromamba/envs/tensorqtl_r/lib/python3.8/site-packages/pandas/core/indexes/base.py:3653\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3652\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 3653\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcasted_key\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3654\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "File \u001b[0;32m~/micromamba/envs/tensorqtl_r/lib/python3.8/site-packages/pandas/_libs/index.pyx:147\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32m~/micromamba/envs/tensorqtl_r/lib/python3.8/site-packages/pandas/_libs/index.pyx:176\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:7080\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:7088\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'cluster_pc_credibleset'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[62], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mpcphenotype_pcvariance\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mcluster_pc_credibleset\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;241m.\u001b[39mstr\u001b[38;5;241m.\u001b[39msplit[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m_pc\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39mstr[\u001b[38;5;241m0\u001b[39m]\n",
      "File \u001b[0;32m~/micromamba/envs/tensorqtl_r/lib/python3.8/site-packages/pandas/core/frame.py:3761\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3759\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcolumns\u001b[38;5;241m.\u001b[39mnlevels \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m1\u001b[39m:\n\u001b[1;32m   3760\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_getitem_multilevel(key)\n\u001b[0;32m-> 3761\u001b[0m indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3762\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_integer(indexer):\n\u001b[1;32m   3763\u001b[0m     indexer \u001b[38;5;241m=\u001b[39m [indexer]\n",
      "File \u001b[0;32m~/micromamba/envs/tensorqtl_r/lib/python3.8/site-packages/pandas/core/indexes/base.py:3655\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3653\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine\u001b[38;5;241m.\u001b[39mget_loc(casted_key)\n\u001b[1;32m   3654\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n\u001b[0;32m-> 3655\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01merr\u001b[39;00m\n\u001b[1;32m   3656\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[1;32m   3657\u001b[0m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[1;32m   3658\u001b[0m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[1;32m   3659\u001b[0m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n\u001b[1;32m   3660\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_indexing_error(key)\n",
      "\u001b[0;31mKeyError\u001b[0m: 'cluster_pc_credibleset'"
     ]
    }
   ],
   "source": [
    "pcphenotype_pcvariance['cluster_pc_credibleset'].str.split['_pc'].str[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variance_weighted_x</th>\n",
       "      <th>cluster_id_x</th>\n",
       "      <th>variant_id_x</th>\n",
       "      <th>cs_id</th>\n",
       "      <th>variance_weighted_y</th>\n",
       "      <th>cluster_id_y</th>\n",
       "      <th>variant_id_y</th>\n",
       "      <th>egene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.381875</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13</td>\n",
       "      <td>chr22_50504763_C_T_b38</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_2</td>\n",
       "      <td>0.246960</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13</td>\n",
       "      <td>chr22_50504763_C_T_b38</td>\n",
       "      <td>ENSG00000025708.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.381875</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13</td>\n",
       "      <td>chr22_50504763_C_T_b38</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_2</td>\n",
       "      <td>0.292988</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13</td>\n",
       "      <td>chr22_50504763_C_T_b38</td>\n",
       "      <td>ENSG00000177989.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.011984</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13</td>\n",
       "      <td>chr22_50504763_C_T_b38</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_2</td>\n",
       "      <td>0.246960</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13</td>\n",
       "      <td>chr22_50504763_C_T_b38</td>\n",
       "      <td>ENSG00000025708.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.011984</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13</td>\n",
       "      <td>chr22_50504763_C_T_b38</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13_2</td>\n",
       "      <td>0.292988</td>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13</td>\n",
       "      <td>chr22_50504763_C_T_b38</td>\n",
       "      <td>ENSG00000177989.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.294142</td>\n",
       "      <td>ENSG00000093000.18_ENSG00000100364.18</td>\n",
       "      <td>chr22_45105543_G_A_b38</td>\n",
       "      <td>ENSG00000093000.18_ENSG00000100364.18_1</td>\n",
       "      <td>0.350770</td>\n",
       "      <td>ENSG00000093000.18_ENSG00000100364.18</td>\n",
       "      <td>chr22_45105543_G_A_b38</td>\n",
       "      <td>ENSG00000093000.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>0.019338</td>\n",
       "      <td>ENSG00000183066.14_ENSG00000198951.11</td>\n",
       "      <td>chr22_41992822_A_G_b38</td>\n",
       "      <td>ENSG00000183066.14_ENSG00000198951.11_1</td>\n",
       "      <td>0.252443</td>\n",
       "      <td>ENSG00000183066.14_ENSG00000198951.11</td>\n",
       "      <td>chr22_41992822_A_G_b38</td>\n",
       "      <td>ENSG00000198951.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>0.268754</td>\n",
       "      <td>ENSG00000243811.8_ENSG00000244509.3</td>\n",
       "      <td>chr22_39023260_C_T_b38</td>\n",
       "      <td>ENSG00000243811.8_ENSG00000244509.3_1</td>\n",
       "      <td>0.230780</td>\n",
       "      <td>ENSG00000243811.8_ENSG00000244509.3</td>\n",
       "      <td>chr22_39023260_C_T_b38</td>\n",
       "      <td>ENSG00000243811.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>0.268754</td>\n",
       "      <td>ENSG00000243811.8_ENSG00000244509.3</td>\n",
       "      <td>chr22_39023260_C_T_b38</td>\n",
       "      <td>ENSG00000243811.8_ENSG00000244509.3_1</td>\n",
       "      <td>0.124338</td>\n",
       "      <td>ENSG00000243811.8_ENSG00000244509.3</td>\n",
       "      <td>chr22_39023260_C_T_b38</td>\n",
       "      <td>ENSG00000244509.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>0.012858</td>\n",
       "      <td>ENSG00000243811.8_ENSG00000244509.3</td>\n",
       "      <td>chr22_39023260_C_T_b38</td>\n",
       "      <td>ENSG00000243811.8_ENSG00000244509.3_1</td>\n",
       "      <td>0.230780</td>\n",
       "      <td>ENSG00000243811.8_ENSG00000244509.3</td>\n",
       "      <td>chr22_39023260_C_T_b38</td>\n",
       "      <td>ENSG00000243811.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>0.012858</td>\n",
       "      <td>ENSG00000243811.8_ENSG00000244509.3</td>\n",
       "      <td>chr22_39023260_C_T_b38</td>\n",
       "      <td>ENSG00000243811.8_ENSG00000244509.3_1</td>\n",
       "      <td>0.124338</td>\n",
       "      <td>ENSG00000243811.8_ENSG00000244509.3</td>\n",
       "      <td>chr22_39023260_C_T_b38</td>\n",
       "      <td>ENSG00000244509.3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>68 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    variance_weighted_x                           cluster_id_x  \\\n",
       "0              0.381875  ENSG00000025708.13_ENSG00000177989.13   \n",
       "1              0.381875  ENSG00000025708.13_ENSG00000177989.13   \n",
       "2              0.011984  ENSG00000025708.13_ENSG00000177989.13   \n",
       "3              0.011984  ENSG00000025708.13_ENSG00000177989.13   \n",
       "4              0.294142  ENSG00000093000.18_ENSG00000100364.18   \n",
       "..                  ...                                    ...   \n",
       "63             0.019338  ENSG00000183066.14_ENSG00000198951.11   \n",
       "64             0.268754    ENSG00000243811.8_ENSG00000244509.3   \n",
       "65             0.268754    ENSG00000243811.8_ENSG00000244509.3   \n",
       "66             0.012858    ENSG00000243811.8_ENSG00000244509.3   \n",
       "67             0.012858    ENSG00000243811.8_ENSG00000244509.3   \n",
       "\n",
       "              variant_id_x                                    cs_id  \\\n",
       "0   chr22_50504763_C_T_b38  ENSG00000025708.13_ENSG00000177989.13_2   \n",
       "1   chr22_50504763_C_T_b38  ENSG00000025708.13_ENSG00000177989.13_2   \n",
       "2   chr22_50504763_C_T_b38  ENSG00000025708.13_ENSG00000177989.13_2   \n",
       "3   chr22_50504763_C_T_b38  ENSG00000025708.13_ENSG00000177989.13_2   \n",
       "4   chr22_45105543_G_A_b38  ENSG00000093000.18_ENSG00000100364.18_1   \n",
       "..                     ...                                      ...   \n",
       "63  chr22_41992822_A_G_b38  ENSG00000183066.14_ENSG00000198951.11_1   \n",
       "64  chr22_39023260_C_T_b38    ENSG00000243811.8_ENSG00000244509.3_1   \n",
       "65  chr22_39023260_C_T_b38    ENSG00000243811.8_ENSG00000244509.3_1   \n",
       "66  chr22_39023260_C_T_b38    ENSG00000243811.8_ENSG00000244509.3_1   \n",
       "67  chr22_39023260_C_T_b38    ENSG00000243811.8_ENSG00000244509.3_1   \n",
       "\n",
       "    variance_weighted_y                           cluster_id_y  \\\n",
       "0              0.246960  ENSG00000025708.13_ENSG00000177989.13   \n",
       "1              0.292988  ENSG00000025708.13_ENSG00000177989.13   \n",
       "2              0.246960  ENSG00000025708.13_ENSG00000177989.13   \n",
       "3              0.292988  ENSG00000025708.13_ENSG00000177989.13   \n",
       "4              0.350770  ENSG00000093000.18_ENSG00000100364.18   \n",
       "..                  ...                                    ...   \n",
       "63             0.252443  ENSG00000183066.14_ENSG00000198951.11   \n",
       "64             0.230780    ENSG00000243811.8_ENSG00000244509.3   \n",
       "65             0.124338    ENSG00000243811.8_ENSG00000244509.3   \n",
       "66             0.230780    ENSG00000243811.8_ENSG00000244509.3   \n",
       "67             0.124338    ENSG00000243811.8_ENSG00000244509.3   \n",
       "\n",
       "              variant_id_y            egene_id  \n",
       "0   chr22_50504763_C_T_b38  ENSG00000025708.13  \n",
       "1   chr22_50504763_C_T_b38  ENSG00000177989.13  \n",
       "2   chr22_50504763_C_T_b38  ENSG00000025708.13  \n",
       "3   chr22_50504763_C_T_b38  ENSG00000177989.13  \n",
       "4   chr22_45105543_G_A_b38  ENSG00000093000.18  \n",
       "..                     ...                 ...  \n",
       "63  chr22_41992822_A_G_b38  ENSG00000198951.11  \n",
       "64  chr22_39023260_C_T_b38   ENSG00000243811.8  \n",
       "65  chr22_39023260_C_T_b38   ENSG00000244509.3  \n",
       "66  chr22_39023260_C_T_b38   ENSG00000243811.8  \n",
       "67  chr22_39023260_C_T_b38   ENSG00000244509.3  \n",
       "\n",
       "[68 rows x 8 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.merge(pcphenotype_pcvariance, pcphenotype_evariance, on = 'cs_id')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I now have a table with a row for the % evar explained for each egene and one with the % pcvar for the pc-overall. I want to further annotate this with relevant info, like the pc-qtl lead variant to egene tss distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load in the gene information (start and strand are what I need)\n",
    "full_gencode=pd.read_csv('/home/klawren/oak/pcqtls/data/references/processed_gencode.v26.GRCh38.genes.gtf', sep='\\t', skiprows=range(6), \n",
    "            header=None, names=['chr', 'dataset', 'type', 'start','end', '.', 'strand', 'na', 'info'])\n",
    "\n",
    "full_gencode = full_gencode[full_gencode['type']=='transcript']\n",
    "full_gencode['transcript_id'] = full_gencode['info'].str.split(';').str[1].str.split('\\\"').str[-2]\n",
    "\n",
    "# add in the start and end info\n",
    "full_gencode['tss_start'] = np.where(full_gencode['strand'] == '+', full_gencode['start'], full_gencode['end'])\n",
    "full_gencode['gene_end'] = np.where(full_gencode['strand'] == '-', full_gencode['start'], full_gencode['end'])\n",
    "\n",
    "# filter to just the transcripts that are in the clusters\n",
    "gene_ids = np.concatenate(overlap_df['cluster_id'].str.split('_'))\n",
    "gid_gencode = full_gencode.set_index('transcript_id').loc[gene_ids]\n",
    "gid_gencode = gid_gencode.drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "row = sum_variance_df.iloc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'egene_id'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m~/micromamba/envs/tensorqtl_r/lib/python3.8/site-packages/pandas/core/indexes/base.py:3653\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3652\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 3653\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcasted_key\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3654\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "File \u001b[0;32m~/micromamba/envs/tensorqtl_r/lib/python3.8/site-packages/pandas/_libs/index.pyx:147\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32m~/micromamba/envs/tensorqtl_r/lib/python3.8/site-packages/pandas/_libs/index.pyx:176\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:7080\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:7088\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'egene_id'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[19], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m egene_tss \u001b[38;5;241m=\u001b[39m gid_gencode\u001b[38;5;241m.\u001b[39mloc[\u001b[43mrow\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43megene_id\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m][\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtss_start\u001b[39m\u001b[38;5;124m'\u001b[39m]\n",
      "File \u001b[0;32m~/micromamba/envs/tensorqtl_r/lib/python3.8/site-packages/pandas/core/series.py:1007\u001b[0m, in \u001b[0;36mSeries.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   1004\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_values[key]\n\u001b[1;32m   1006\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m key_is_scalar:\n\u001b[0;32m-> 1007\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_value\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1009\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_hashable(key):\n\u001b[1;32m   1010\u001b[0m     \u001b[38;5;66;03m# Otherwise index.get_value will raise InvalidIndexError\u001b[39;00m\n\u001b[1;32m   1011\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m   1012\u001b[0m         \u001b[38;5;66;03m# For labels that don't resolve as scalars like tuples and frozensets\u001b[39;00m\n",
      "File \u001b[0;32m~/micromamba/envs/tensorqtl_r/lib/python3.8/site-packages/pandas/core/series.py:1116\u001b[0m, in \u001b[0;36mSeries._get_value\u001b[0;34m(self, label, takeable)\u001b[0m\n\u001b[1;32m   1113\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_values[label]\n\u001b[1;32m   1115\u001b[0m \u001b[38;5;66;03m# Similar to Index.get_value, but we do not fall back to positional\u001b[39;00m\n\u001b[0;32m-> 1116\u001b[0m loc \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mindex\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mlabel\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1118\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_integer(loc):\n\u001b[1;32m   1119\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_values[loc]\n",
      "File \u001b[0;32m~/micromamba/envs/tensorqtl_r/lib/python3.8/site-packages/pandas/core/indexes/base.py:3655\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3653\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine\u001b[38;5;241m.\u001b[39mget_loc(casted_key)\n\u001b[1;32m   3654\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n\u001b[0;32m-> 3655\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01merr\u001b[39;00m\n\u001b[1;32m   3656\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[1;32m   3657\u001b[0m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[1;32m   3658\u001b[0m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[1;32m   3659\u001b[0m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n\u001b[1;32m   3660\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_indexing_error(key)\n",
      "\u001b[0;31mKeyError\u001b[0m: 'egene_id'"
     ]
    }
   ],
   "source": [
    "egene_tss = gid_gencode.loc[row['egene_id']]['tss_start']\n",
    "lead"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
