{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### make smaller eqtl and gwas to work with "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_clusters(TISSUE, CHROM):\n",
    "    # Logic to generate a list of sample names\n",
    "    return open(coloc_output_dir + '{TISSUE}/temp/{TISSUE}.{CHROM}.cluster_list.txt').read().strip().split(\"\\n\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_path = '/home/klawren/oak/pcqtls/output/proteincoding_main/clusters/Adipose_Subcutaneous_clusters_all_chr.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "        import pandas as pd\n",
    "        cluster_df =  pd.read_csv(cluster_path,index_col=0)\n",
    "        for idx, row in cluster_df.iterrows():\n",
    "            cluster_df.loc[idx, 'cluster_id']  = '_'.join([*sorted(row['Transcripts'].split(','))])\n",
    "        output_head = '/home/klawren/oak/pcqtls/output/'\n",
    "        tissue_id = 'Adipose_Subcutaneous'\n",
    "        for chrom in cluster_df['Chromosome'].unique():\n",
    "            cluster_df[cluster_df['Chromosome']==chrom]['cluster_id'].to_csv(f'{output_head}/temp/{tissue_id}.chr{chrom}.cluster_list.txt', index=False, header=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "eqtl = pd.read_parquet('output/proteincoding_main/control_eqtl/Adipose_Subcutaneous/Adipose_Subcutaneous.v8.cluster_genes.cis_qtl_pairs.chr22.parquet')\n",
    "cluster_list = pd.read_csv(f'{output_head}/temp/{tissue_id}.chr22.cluster_list.txt', header=None, names=['cluster_id'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000099937.10_ENSG00000099940.11_ENSG0000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000099974.7_ENSG00000133433.10_ENSG00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000100342.20_ENSG00000128335.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000025708.13_ENSG00000177989.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000100429.17_ENSG00000188130.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSG00000075240.16_ENSG00000100422.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSG00000243811.8_ENSG00000244509.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSG00000100221.10_ENSG00000100226.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSG00000100151.15_ENSG00000100156.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSG00000100139.13_ENSG00000128346.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENSG00000100060.17_ENSG00000100065.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENSG00000128309.16_ENSG00000128311.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENSG00000128313.2_ENSG00000221963.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENSG00000100234.11_ENSG00000185666.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENSG00000099992.15_ENSG00000239282.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENSG00000100276.9_ENSG00000100280.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENSG00000099991.17_ENSG00000099994.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ENSG00000099953.9_ENSG00000099956.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ENSG00000184058.12_ENSG00000215012.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ENSG00000100033.16_ENSG00000183628.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ENSG00000183785.14_ENSG00000184979.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ENSG00000015475.18_ENSG00000099968.17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ENSG00000100312.10_ENSG00000251322.7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           cluster_id\n",
       "0   ENSG00000099937.10_ENSG00000099940.11_ENSG0000...\n",
       "1   ENSG00000099974.7_ENSG00000133433.10_ENSG00000...\n",
       "2               ENSG00000100342.20_ENSG00000128335.13\n",
       "3               ENSG00000025708.13_ENSG00000177989.13\n",
       "4               ENSG00000100429.17_ENSG00000188130.13\n",
       "5               ENSG00000075240.16_ENSG00000100422.13\n",
       "6                 ENSG00000243811.8_ENSG00000244509.3\n",
       "7               ENSG00000100221.10_ENSG00000100226.15\n",
       "8               ENSG00000100151.15_ENSG00000100156.10\n",
       "9               ENSG00000100139.13_ENSG00000128346.10\n",
       "10              ENSG00000100060.17_ENSG00000100065.14\n",
       "11              ENSG00000128309.16_ENSG00000128311.13\n",
       "12                ENSG00000128313.2_ENSG00000221963.5\n",
       "13              ENSG00000100234.11_ENSG00000185666.14\n",
       "14               ENSG00000099992.15_ENSG00000239282.7\n",
       "15               ENSG00000100276.9_ENSG00000100280.16\n",
       "16              ENSG00000099991.17_ENSG00000099994.10\n",
       "17               ENSG00000099953.9_ENSG00000099956.18\n",
       "18               ENSG00000184058.12_ENSG00000215012.8\n",
       "19              ENSG00000100033.16_ENSG00000183628.12\n",
       "20               ENSG00000183785.14_ENSG00000184979.9\n",
       "21              ENSG00000015475.18_ENSG00000099968.17\n",
       "22               ENSG00000100312.10_ENSG00000251322.7"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cluster_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "eqtl['cluster_id'] = eqtl['phenotype_id'].str.split('_e').str[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "for idx, row in cluster_list.iterrows():\n",
    "    snp_list = eqtl[eqtl['cluster_id'] == row['cluster_id']]['variant_id']\n",
    "    snp_list.to_csv('output/temp/snp_list.csv', index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for idx, row in eqtl.iterrows():\n",
    "       eqtl.loc[idx, 'cluster_id']  = row['Transcripts'].split(','))])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "sub_eqtl = eqtl[eqtl['phenotype_id'] == eqtl['phenotype_id'].iloc[0]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "sub_eqtl.to_csv('output/temp/test_eqtl.csv', index=None, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "gwas = pd.read_csv('/oak/stanford/groups/smontgom/shared/gwas_summary_stats/barbeira_gtex_imputed/imputed_gwas_hg38_1.1/imputed_UKB_50_Standing_height.txt.gz', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>panel_variant_id</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>effect_allele</th>\n",
       "      <th>non_effect_allele</th>\n",
       "      <th>current_build</th>\n",
       "      <th>frequency</th>\n",
       "      <th>sample_size</th>\n",
       "      <th>zscore</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>effect_size</th>\n",
       "      <th>standard_error</th>\n",
       "      <th>imputation_status</th>\n",
       "      <th>n_cases</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs554008981</td>\n",
       "      <td>chr1_13550_G_A_b38</td>\n",
       "      <td>chr1</td>\n",
       "      <td>13550</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>hg38</td>\n",
       "      <td>0.017316</td>\n",
       "      <td>336474</td>\n",
       "      <td>-2.291993</td>\n",
       "      <td>0.021906</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>imputed</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs201055865</td>\n",
       "      <td>chr1_14671_G_C_b38</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14671</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>hg38</td>\n",
       "      <td>0.012987</td>\n",
       "      <td>336474</td>\n",
       "      <td>-0.955919</td>\n",
       "      <td>0.339113</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>imputed</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs201327123</td>\n",
       "      <td>chr1_14677_G_A_b38</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14677</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>hg38</td>\n",
       "      <td>0.053391</td>\n",
       "      <td>336474</td>\n",
       "      <td>1.495331</td>\n",
       "      <td>0.134828</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>imputed</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs199856693</td>\n",
       "      <td>chr1_14933_G_A_b38</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14933</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>hg38</td>\n",
       "      <td>0.011544</td>\n",
       "      <td>336474</td>\n",
       "      <td>-0.461158</td>\n",
       "      <td>0.644685</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>imputed</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs62636368</td>\n",
       "      <td>chr1_16841_G_T_b38</td>\n",
       "      <td>chr1</td>\n",
       "      <td>16841</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>hg38</td>\n",
       "      <td>0.045455</td>\n",
       "      <td>336474</td>\n",
       "      <td>0.793825</td>\n",
       "      <td>0.427297</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>imputed</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    variant_id    panel_variant_id chromosome  position effect_allele  \\\n",
       "0  rs554008981  chr1_13550_G_A_b38       chr1     13550             A   \n",
       "1  rs201055865  chr1_14671_G_C_b38       chr1     14671             C   \n",
       "2  rs201327123  chr1_14677_G_A_b38       chr1     14677             A   \n",
       "3  rs199856693  chr1_14933_G_A_b38       chr1     14933             A   \n",
       "4   rs62636368  chr1_16841_G_T_b38       chr1     16841             T   \n",
       "\n",
       "  non_effect_allele current_build  frequency  sample_size    zscore    pvalue  \\\n",
       "0                 G          hg38   0.017316       336474 -2.291993  0.021906   \n",
       "1                 G          hg38   0.012987       336474 -0.955919  0.339113   \n",
       "2                 G          hg38   0.053391       336474  1.495331  0.134828   \n",
       "3                 G          hg38   0.011544       336474 -0.461158  0.644685   \n",
       "4                 G          hg38   0.045455       336474  0.793825  0.427297   \n",
       "\n",
       "   effect_size  standard_error imputation_status  n_cases  \n",
       "0          NaN             NaN           imputed      NaN  \n",
       "1          NaN             NaN           imputed      NaN  \n",
       "2          NaN             NaN           imputed      NaN  \n",
       "3          NaN             NaN           imputed      NaN  \n",
       "4          NaN             NaN           imputed      NaN  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "eqtl['position'] = eqtl['variant_id'].str.split('_').str[1].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "max_position = sub_eqtl['variant_id'].str.split('_').str[1].astype(int).max()\n",
    "min_position = sub_eqtl['variant_id'].str.split('_').str[1].astype(int).min()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "gwas = gwas[gwas['chromosome'] == 'chr22']\n",
    "gwas.fillna('NA', inplace=True)\n",
    "sub_gwas = gwas[(gwas['position'].astype(int) > min_position) & (gwas['position'].astype(int) < max_position)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_list = sub_eqtl[sub_eqtl['variant_id'].isin(sub_gwas['panel_variant_id'])]['variant_id']\n",
    "snp_list.to_csv('output/temp/snp_list.csv', index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "sub_gwas.to_csv('output/temp/test_gwas.csv', index=None, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "ld_matrix = pd.read_csv('output/temp/test.ld', sep='\\t', header=None, names=snp_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "ld_matrix.set_index(snp_list, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "178"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ld_matrix.isna().sum().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "chr22_16956420_C_T_b38"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "variant_id\n",
       "chr22_16956420_C_T_b38   NaN\n",
       "Name: chr22_16954390_G_A_b38, dtype: float64"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ld_matrix['chr22_16954390_G_A_b38'][ld_matrix['chr22_16954390_G_A_b38'].isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr22_16628911_CTTG_C_b38</th>\n",
       "      <th>chr22_16629375_T_A_b38</th>\n",
       "      <th>chr22_16629512_C_G_b38</th>\n",
       "      <th>chr22_16629863_T_G_b38</th>\n",
       "      <th>chr22_16630596_A_T_b38</th>\n",
       "      <th>chr22_16631452_C_T_b38</th>\n",
       "      <th>chr22_16631637_C_CCAT_b38</th>\n",
       "      <th>chr22_16631746_C_T_b38</th>\n",
       "      <th>chr22_16632220_A_G_b38</th>\n",
       "      <th>chr22_16632936_G_T_b38</th>\n",
       "      <th>...</th>\n",
       "      <th>chr22_18637212_A_G_b38</th>\n",
       "      <th>chr22_18637251_G_T_b38</th>\n",
       "      <th>chr22_18638681_C_CT_b38</th>\n",
       "      <th>chr22_18658980_A_G_b38</th>\n",
       "      <th>chr22_18659236_T_C_b38</th>\n",
       "      <th>chr22_18659263_C_G_b38</th>\n",
       "      <th>chr22_18727775_T_A_b38</th>\n",
       "      <th>chr22_18727816_T_C_b38</th>\n",
       "      <th>chr22_18727918_G_C_b38</th>\n",
       "      <th>chr22_18727946_C_T_b38</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>variant_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr22_16954390_G_A_b38</th>\n",
       "      <td>0.000268</td>\n",
       "      <td>0.001199</td>\n",
       "      <td>0.000510</td>\n",
       "      <td>0.000204</td>\n",
       "      <td>0.001530</td>\n",
       "      <td>0.000207</td>\n",
       "      <td>0.000108</td>\n",
       "      <td>0.000928</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>6.591740e-04</td>\n",
       "      <td>...</td>\n",
       "      <td>0.003536</td>\n",
       "      <td>0.001597</td>\n",
       "      <td>0.000267</td>\n",
       "      <td>0.000027</td>\n",
       "      <td>0.000663</td>\n",
       "      <td>0.000447</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.000829</td>\n",
       "      <td>0.001124</td>\n",
       "      <td>0.001018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr22_16955693_G_A_b38</th>\n",
       "      <td>0.000268</td>\n",
       "      <td>0.001199</td>\n",
       "      <td>0.000510</td>\n",
       "      <td>0.000204</td>\n",
       "      <td>0.001530</td>\n",
       "      <td>0.000207</td>\n",
       "      <td>0.000108</td>\n",
       "      <td>0.000928</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>6.591740e-04</td>\n",
       "      <td>...</td>\n",
       "      <td>0.003536</td>\n",
       "      <td>0.001597</td>\n",
       "      <td>0.000267</td>\n",
       "      <td>0.000027</td>\n",
       "      <td>0.000663</td>\n",
       "      <td>0.000447</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.000829</td>\n",
       "      <td>0.001124</td>\n",
       "      <td>0.001018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr22_16955737_C_T_b38</th>\n",
       "      <td>0.000268</td>\n",
       "      <td>0.001199</td>\n",
       "      <td>0.000510</td>\n",
       "      <td>0.000204</td>\n",
       "      <td>0.001530</td>\n",
       "      <td>0.000207</td>\n",
       "      <td>0.000108</td>\n",
       "      <td>0.000928</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>6.591740e-04</td>\n",
       "      <td>...</td>\n",
       "      <td>0.003536</td>\n",
       "      <td>0.001597</td>\n",
       "      <td>0.000267</td>\n",
       "      <td>0.000027</td>\n",
       "      <td>0.000663</td>\n",
       "      <td>0.000447</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.000829</td>\n",
       "      <td>0.001124</td>\n",
       "      <td>0.001018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr22_16956420_C_T_b38</th>\n",
       "      <td>0.000164</td>\n",
       "      <td>0.000185</td>\n",
       "      <td>0.000485</td>\n",
       "      <td>0.003975</td>\n",
       "      <td>0.001848</td>\n",
       "      <td>0.000495</td>\n",
       "      <td>0.000329</td>\n",
       "      <td>0.000355</td>\n",
       "      <td>0.000102</td>\n",
       "      <td>4.488370e-07</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>0.000235</td>\n",
       "      <td>0.000017</td>\n",
       "      <td>0.001307</td>\n",
       "      <td>0.000081</td>\n",
       "      <td>0.000108</td>\n",
       "      <td>0.000302</td>\n",
       "      <td>0.000157</td>\n",
       "      <td>0.003610</td>\n",
       "      <td>0.003775</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr22_16956485_G_C_b38</th>\n",
       "      <td>0.000268</td>\n",
       "      <td>0.001199</td>\n",
       "      <td>0.000510</td>\n",
       "      <td>0.000204</td>\n",
       "      <td>0.001530</td>\n",
       "      <td>0.000207</td>\n",
       "      <td>0.000108</td>\n",
       "      <td>0.000928</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>6.591740e-04</td>\n",
       "      <td>...</td>\n",
       "      <td>0.003536</td>\n",
       "      <td>0.001597</td>\n",
       "      <td>0.000267</td>\n",
       "      <td>0.000027</td>\n",
       "      <td>0.000663</td>\n",
       "      <td>0.000447</td>\n",
       "      <td>0.000008</td>\n",
       "      <td>0.000829</td>\n",
       "      <td>0.001124</td>\n",
       "      <td>0.001018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr22_18132365_C_T_b38</th>\n",
       "      <td>0.002062</td>\n",
       "      <td>0.000443</td>\n",
       "      <td>0.000129</td>\n",
       "      <td>0.000086</td>\n",
       "      <td>0.000549</td>\n",
       "      <td>0.001831</td>\n",
       "      <td>0.000156</td>\n",
       "      <td>0.000268</td>\n",
       "      <td>0.000092</td>\n",
       "      <td>7.427620e-04</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000027</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>0.000918</td>\n",
       "      <td>0.000223</td>\n",
       "      <td>0.002478</td>\n",
       "      <td>0.005141</td>\n",
       "      <td>0.000162</td>\n",
       "      <td>0.000998</td>\n",
       "      <td>0.003670</td>\n",
       "      <td>0.010987</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr22_18135190_T_TG_b38</th>\n",
       "      <td>0.002993</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000132</td>\n",
       "      <td>0.000179</td>\n",
       "      <td>0.000092</td>\n",
       "      <td>0.002614</td>\n",
       "      <td>0.000213</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>0.001065</td>\n",
       "      <td>7.887450e-05</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000962</td>\n",
       "      <td>0.000439</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.002364</td>\n",
       "      <td>0.002452</td>\n",
       "      <td>0.002707</td>\n",
       "      <td>0.000908</td>\n",
       "      <td>0.000324</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.001248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr22_18135974_T_C_b38</th>\n",
       "      <td>0.002993</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000132</td>\n",
       "      <td>0.000179</td>\n",
       "      <td>0.000092</td>\n",
       "      <td>0.002614</td>\n",
       "      <td>0.000213</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>0.001065</td>\n",
       "      <td>7.887450e-05</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000962</td>\n",
       "      <td>0.000439</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.002364</td>\n",
       "      <td>0.002452</td>\n",
       "      <td>0.002707</td>\n",
       "      <td>0.000908</td>\n",
       "      <td>0.000324</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.001248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr22_18136084_G_T_b38</th>\n",
       "      <td>0.002993</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.000132</td>\n",
       "      <td>0.000179</td>\n",
       "      <td>0.000092</td>\n",
       "      <td>0.002614</td>\n",
       "      <td>0.000213</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>0.001065</td>\n",
       "      <td>7.887450e-05</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000962</td>\n",
       "      <td>0.000439</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.002364</td>\n",
       "      <td>0.002452</td>\n",
       "      <td>0.002707</td>\n",
       "      <td>0.000908</td>\n",
       "      <td>0.000324</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>0.001248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr22_18141296_C_G_b38</th>\n",
       "      <td>0.002033</td>\n",
       "      <td>0.000740</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>0.000200</td>\n",
       "      <td>0.000050</td>\n",
       "      <td>0.001839</td>\n",
       "      <td>0.000017</td>\n",
       "      <td>0.000545</td>\n",
       "      <td>0.001090</td>\n",
       "      <td>1.270370e-03</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000102</td>\n",
       "      <td>0.001100</td>\n",
       "      <td>0.000064</td>\n",
       "      <td>0.000124</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.009406</td>\n",
       "      <td>0.000180</td>\n",
       "      <td>0.004044</td>\n",
       "      <td>0.009620</td>\n",
       "      <td>0.023255</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>94 rows Ã— 6408 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         chr22_16628911_CTTG_C_b38  chr22_16629375_T_A_b38  \\\n",
       "variant_id                                                                   \n",
       "chr22_16954390_G_A_b38                    0.000268                0.001199   \n",
       "chr22_16955693_G_A_b38                    0.000268                0.001199   \n",
       "chr22_16955737_C_T_b38                    0.000268                0.001199   \n",
       "chr22_16956420_C_T_b38                    0.000164                0.000185   \n",
       "chr22_16956485_G_C_b38                    0.000268                0.001199   \n",
       "...                                            ...                     ...   \n",
       "chr22_18132365_C_T_b38                    0.002062                0.000443   \n",
       "chr22_18135190_T_TG_b38                   0.002993                0.000002   \n",
       "chr22_18135974_T_C_b38                    0.002993                0.000002   \n",
       "chr22_18136084_G_T_b38                    0.002993                0.000002   \n",
       "chr22_18141296_C_G_b38                    0.002033                0.000740   \n",
       "\n",
       "                         chr22_16629512_C_G_b38  chr22_16629863_T_G_b38  \\\n",
       "variant_id                                                                \n",
       "chr22_16954390_G_A_b38                 0.000510                0.000204   \n",
       "chr22_16955693_G_A_b38                 0.000510                0.000204   \n",
       "chr22_16955737_C_T_b38                 0.000510                0.000204   \n",
       "chr22_16956420_C_T_b38                 0.000485                0.003975   \n",
       "chr22_16956485_G_C_b38                 0.000510                0.000204   \n",
       "...                                         ...                     ...   \n",
       "chr22_18132365_C_T_b38                 0.000129                0.000086   \n",
       "chr22_18135190_T_TG_b38                0.000132                0.000179   \n",
       "chr22_18135974_T_C_b38                 0.000132                0.000179   \n",
       "chr22_18136084_G_T_b38                 0.000132                0.000179   \n",
       "chr22_18141296_C_G_b38                 0.000042                0.000200   \n",
       "\n",
       "                         chr22_16630596_A_T_b38  chr22_16631452_C_T_b38  \\\n",
       "variant_id                                                                \n",
       "chr22_16954390_G_A_b38                 0.001530                0.000207   \n",
       "chr22_16955693_G_A_b38                 0.001530                0.000207   \n",
       "chr22_16955737_C_T_b38                 0.001530                0.000207   \n",
       "chr22_16956420_C_T_b38                 0.001848                0.000495   \n",
       "chr22_16956485_G_C_b38                 0.001530                0.000207   \n",
       "...                                         ...                     ...   \n",
       "chr22_18132365_C_T_b38                 0.000549                0.001831   \n",
       "chr22_18135190_T_TG_b38                0.000092                0.002614   \n",
       "chr22_18135974_T_C_b38                 0.000092                0.002614   \n",
       "chr22_18136084_G_T_b38                 0.000092                0.002614   \n",
       "chr22_18141296_C_G_b38                 0.000050                0.001839   \n",
       "\n",
       "                         chr22_16631637_C_CCAT_b38  chr22_16631746_C_T_b38  \\\n",
       "variant_id                                                                   \n",
       "chr22_16954390_G_A_b38                    0.000108                0.000928   \n",
       "chr22_16955693_G_A_b38                    0.000108                0.000928   \n",
       "chr22_16955737_C_T_b38                    0.000108                0.000928   \n",
       "chr22_16956420_C_T_b38                    0.000329                0.000355   \n",
       "chr22_16956485_G_C_b38                    0.000108                0.000928   \n",
       "...                                            ...                     ...   \n",
       "chr22_18132365_C_T_b38                    0.000156                0.000268   \n",
       "chr22_18135190_T_TG_b38                   0.000213                0.000022   \n",
       "chr22_18135974_T_C_b38                    0.000213                0.000022   \n",
       "chr22_18136084_G_T_b38                    0.000213                0.000022   \n",
       "chr22_18141296_C_G_b38                    0.000017                0.000545   \n",
       "\n",
       "                         chr22_16632220_A_G_b38  chr22_16632936_G_T_b38  ...  \\\n",
       "variant_id                                                               ...   \n",
       "chr22_16954390_G_A_b38                 0.000003            6.591740e-04  ...   \n",
       "chr22_16955693_G_A_b38                 0.000003            6.591740e-04  ...   \n",
       "chr22_16955737_C_T_b38                 0.000003            6.591740e-04  ...   \n",
       "chr22_16956420_C_T_b38                 0.000102            4.488370e-07  ...   \n",
       "chr22_16956485_G_C_b38                 0.000003            6.591740e-04  ...   \n",
       "...                                         ...                     ...  ...   \n",
       "chr22_18132365_C_T_b38                 0.000092            7.427620e-04  ...   \n",
       "chr22_18135190_T_TG_b38                0.001065            7.887450e-05  ...   \n",
       "chr22_18135974_T_C_b38                 0.001065            7.887450e-05  ...   \n",
       "chr22_18136084_G_T_b38                 0.001065            7.887450e-05  ...   \n",
       "chr22_18141296_C_G_b38                 0.001090            1.270370e-03  ...   \n",
       "\n",
       "                         chr22_18637212_A_G_b38  chr22_18637251_G_T_b38  \\\n",
       "variant_id                                                                \n",
       "chr22_16954390_G_A_b38                 0.003536                0.001597   \n",
       "chr22_16955693_G_A_b38                 0.003536                0.001597   \n",
       "chr22_16955737_C_T_b38                 0.003536                0.001597   \n",
       "chr22_16956420_C_T_b38                 0.000100                0.000235   \n",
       "chr22_16956485_G_C_b38                 0.003536                0.001597   \n",
       "...                                         ...                     ...   \n",
       "chr22_18132365_C_T_b38                 0.000027                0.000009   \n",
       "chr22_18135190_T_TG_b38                0.000962                0.000439   \n",
       "chr22_18135974_T_C_b38                 0.000962                0.000439   \n",
       "chr22_18136084_G_T_b38                 0.000962                0.000439   \n",
       "chr22_18141296_C_G_b38                 0.000102                0.001100   \n",
       "\n",
       "                         chr22_18638681_C_CT_b38  chr22_18658980_A_G_b38  \\\n",
       "variant_id                                                                 \n",
       "chr22_16954390_G_A_b38                  0.000267                0.000027   \n",
       "chr22_16955693_G_A_b38                  0.000267                0.000027   \n",
       "chr22_16955737_C_T_b38                  0.000267                0.000027   \n",
       "chr22_16956420_C_T_b38                  0.000017                0.001307   \n",
       "chr22_16956485_G_C_b38                  0.000267                0.000027   \n",
       "...                                          ...                     ...   \n",
       "chr22_18132365_C_T_b38                  0.000918                0.000223   \n",
       "chr22_18135190_T_TG_b38                 0.000015                0.002364   \n",
       "chr22_18135974_T_C_b38                  0.000015                0.002364   \n",
       "chr22_18136084_G_T_b38                  0.000015                0.002364   \n",
       "chr22_18141296_C_G_b38                  0.000064                0.000124   \n",
       "\n",
       "                         chr22_18659236_T_C_b38  chr22_18659263_C_G_b38  \\\n",
       "variant_id                                                                \n",
       "chr22_16954390_G_A_b38                 0.000663                0.000447   \n",
       "chr22_16955693_G_A_b38                 0.000663                0.000447   \n",
       "chr22_16955737_C_T_b38                 0.000663                0.000447   \n",
       "chr22_16956420_C_T_b38                 0.000081                0.000108   \n",
       "chr22_16956485_G_C_b38                 0.000663                0.000447   \n",
       "...                                         ...                     ...   \n",
       "chr22_18132365_C_T_b38                 0.002478                0.005141   \n",
       "chr22_18135190_T_TG_b38                0.002452                0.002707   \n",
       "chr22_18135974_T_C_b38                 0.002452                0.002707   \n",
       "chr22_18136084_G_T_b38                 0.002452                0.002707   \n",
       "chr22_18141296_C_G_b38                 0.000026                0.009406   \n",
       "\n",
       "                         chr22_18727775_T_A_b38  chr22_18727816_T_C_b38  \\\n",
       "variant_id                                                                \n",
       "chr22_16954390_G_A_b38                 0.000008                0.000829   \n",
       "chr22_16955693_G_A_b38                 0.000008                0.000829   \n",
       "chr22_16955737_C_T_b38                 0.000008                0.000829   \n",
       "chr22_16956420_C_T_b38                 0.000302                0.000157   \n",
       "chr22_16956485_G_C_b38                 0.000008                0.000829   \n",
       "...                                         ...                     ...   \n",
       "chr22_18132365_C_T_b38                 0.000162                0.000998   \n",
       "chr22_18135190_T_TG_b38                0.000908                0.000324   \n",
       "chr22_18135974_T_C_b38                 0.000908                0.000324   \n",
       "chr22_18136084_G_T_b38                 0.000908                0.000324   \n",
       "chr22_18141296_C_G_b38                 0.000180                0.004044   \n",
       "\n",
       "                         chr22_18727918_G_C_b38  chr22_18727946_C_T_b38  \n",
       "variant_id                                                               \n",
       "chr22_16954390_G_A_b38                 0.001124                0.001018  \n",
       "chr22_16955693_G_A_b38                 0.001124                0.001018  \n",
       "chr22_16955737_C_T_b38                 0.001124                0.001018  \n",
       "chr22_16956420_C_T_b38                 0.003610                0.003775  \n",
       "chr22_16956485_G_C_b38                 0.001124                0.001018  \n",
       "...                                         ...                     ...  \n",
       "chr22_18132365_C_T_b38                 0.003670                0.010987  \n",
       "chr22_18135190_T_TG_b38                0.000018                0.001248  \n",
       "chr22_18135974_T_C_b38                 0.000018                0.001248  \n",
       "chr22_18136084_G_T_b38                 0.000018                0.001248  \n",
       "chr22_18141296_C_G_b38                 0.009620                0.023255  \n",
       "\n",
       "[94 rows x 6408 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ld_matrix[ld_matrix.isna().sum()>0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>start_distance</th>\n",
       "      <th>end_distance</th>\n",
       "      <th>af</th>\n",
       "      <th>ma_samples</th>\n",
       "      <th>ma_count</th>\n",
       "      <th>pval_nominal</th>\n",
       "      <th>slope</th>\n",
       "      <th>slope_se</th>\n",
       "      <th>position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1331</th>\n",
       "      <td>ENSG00000015475.18_ENSG00000099968.17_e_ENSG00...</td>\n",
       "      <td>chr22_16956420_C_T_b38</td>\n",
       "      <td>-672435</td>\n",
       "      <td>-818350</td>\n",
       "      <td>0.374322</td>\n",
       "      <td>373</td>\n",
       "      <td>434</td>\n",
       "      <td>0.391571</td>\n",
       "      <td>-0.002587</td>\n",
       "      <td>0.003017</td>\n",
       "      <td>16956420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8819</th>\n",
       "      <td>ENSG00000015475.18_ENSG00000099968.17_e_ENSG00...</td>\n",
       "      <td>chr22_16956420_C_T_b38</td>\n",
       "      <td>-672435</td>\n",
       "      <td>-818350</td>\n",
       "      <td>0.374322</td>\n",
       "      <td>373</td>\n",
       "      <td>434</td>\n",
       "      <td>0.027248</td>\n",
       "      <td>0.006652</td>\n",
       "      <td>0.003004</td>\n",
       "      <td>16956420</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           phenotype_id  \\\n",
       "1331  ENSG00000015475.18_ENSG00000099968.17_e_ENSG00...   \n",
       "8819  ENSG00000015475.18_ENSG00000099968.17_e_ENSG00...   \n",
       "\n",
       "                  variant_id  start_distance  end_distance        af  \\\n",
       "1331  chr22_16956420_C_T_b38         -672435       -818350  0.374322   \n",
       "8819  chr22_16956420_C_T_b38         -672435       -818350  0.374322   \n",
       "\n",
       "      ma_samples  ma_count  pval_nominal     slope  slope_se  position  \n",
       "1331         373       434      0.391571 -0.002587  0.003017  16956420  \n",
       "8819         373       434      0.027248  0.006652  0.003004  16956420  "
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eqtl[eqtl['variant_id']=='chr22_16956420_C_T_b38']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
